// @fluentui/theme/lib/motion/AnimationStyles.js
import { keyframes as e } from "../../../merge-styles/lib/keyframes.js";
var n = "cubic-bezier(.1,.9,.2,1)", r = "cubic-bezier(.1,.25,.75,.9)", _ = "0.167s", l = "0.267s", a = "0.367s", O = "0.467s", c = /* @__PURE__ */ e({
  from: { opacity: 0 },
  to: { opacity: 1 }
}), o = /* @__PURE__ */ e({
  from: { opacity: 1 },
  to: { opacity: 0, visibility: "hidden" }
}), L = /* @__PURE__ */ i(-10), m = /* @__PURE__ */ i(-20), D = /* @__PURE__ */ i(-40), T = /* @__PURE__ */ i(-400), u = /* @__PURE__ */ i(10), S = /* @__PURE__ */ i(20), N = /* @__PURE__ */ i(40), U = /* @__PURE__ */ i(400), R = /* @__PURE__ */ d(10), p = /* @__PURE__ */ d(20), A = /* @__PURE__ */ d(-10), g = /* @__PURE__ */ d(-20), F = /* @__PURE__ */ I(10), h = /* @__PURE__ */ I(20), G = /* @__PURE__ */ I(40), H = /* @__PURE__ */ I(400), y = /* @__PURE__ */ I(-10), b = /* @__PURE__ */ I(-20), w = /* @__PURE__ */ I(-40), C = /* @__PURE__ */ I(-400), P = /* @__PURE__ */ f(-10), W = /* @__PURE__ */ f(-20), x = /* @__PURE__ */ f(10), Z = /* @__PURE__ */ f(20), z = /* @__PURE__ */ e({
  from: { transform: "scale3d(.98,.98,1)" },
  to: { transform: "scale3d(1,1,1)" }
}), X = /* @__PURE__ */ e({
  from: { transform: "scale3d(1,1,1)" },
  to: { transform: "scale3d(.98,.98,1)" }
}), Y = /* @__PURE__ */ e({
  from: { transform: "scale3d(1.03,1.03,1)" },
  to: { transform: "scale3d(1,1,1)" }
}), k = /* @__PURE__ */ e({
  from: { transform: "scale3d(1,1,1)" },
  to: { transform: "scale3d(1.03,1.03,1)" }
}), M = /* @__PURE__ */ e({
  from: { transform: "rotateZ(0deg)" },
  to: { transform: "rotateZ(90deg)" }
}), j = /* @__PURE__ */ e({
  from: { transform: "rotateZ(0deg)" },
  to: { transform: "rotateZ(-90deg)" }
}), B = {
  slideRightIn10: /* @__PURE__ */ t("".concat(c, ",").concat(L), a, n),
  slideRightIn20: /* @__PURE__ */ t("".concat(c, ",").concat(m), a, n),
  slideRightIn40: /* @__PURE__ */ t("".concat(c, ",").concat(D), a, n),
  slideRightIn400: /* @__PURE__ */ t("".concat(c, ",").concat(T), a, n),
  slideLeftIn10: /* @__PURE__ */ t("".concat(c, ",").concat(u), a, n),
  slideLeftIn20: /* @__PURE__ */ t("".concat(c, ",").concat(S), a, n),
  slideLeftIn40: /* @__PURE__ */ t("".concat(c, ",").concat(N), a, n),
  slideLeftIn400: /* @__PURE__ */ t("".concat(c, ",").concat(U), a, n),
  slideUpIn10: /* @__PURE__ */ t("".concat(c, ",").concat(R), a, n),
  slideUpIn20: /* @__PURE__ */ t("".concat(c, ",").concat(p), a, n),
  slideDownIn10: /* @__PURE__ */ t("".concat(c, ",").concat(A), a, n),
  slideDownIn20: /* @__PURE__ */ t("".concat(c, ",").concat(g), a, n),
  slideRightOut10: /* @__PURE__ */ t("".concat(o, ",").concat(F), a, n),
  slideRightOut20: /* @__PURE__ */ t("".concat(o, ",").concat(h), a, n),
  slideRightOut40: /* @__PURE__ */ t("".concat(o, ",").concat(G), a, n),
  slideRightOut400: /* @__PURE__ */ t("".concat(o, ",").concat(H), a, n),
  slideLeftOut10: /* @__PURE__ */ t("".concat(o, ",").concat(y), a, n),
  slideLeftOut20: /* @__PURE__ */ t("".concat(o, ",").concat(b), a, n),
  slideLeftOut40: /* @__PURE__ */ t("".concat(o, ",").concat(w), a, n),
  slideLeftOut400: /* @__PURE__ */ t("".concat(o, ",").concat(C), a, n),
  slideUpOut10: /* @__PURE__ */ t("".concat(o, ",").concat(P), a, n),
  slideUpOut20: /* @__PURE__ */ t("".concat(o, ",").concat(W), a, n),
  slideDownOut10: /* @__PURE__ */ t("".concat(o, ",").concat(x), a, n),
  slideDownOut20: /* @__PURE__ */ t("".concat(o, ",").concat(Z), a, n),
  scaleUpIn100: /* @__PURE__ */ t("".concat(c, ",").concat(z), a, n),
  scaleDownIn100: /* @__PURE__ */ t("".concat(c, ",").concat(Y), a, n),
  scaleUpOut103: /* @__PURE__ */ t("".concat(o, ",").concat(k), _, r),
  scaleDownOut98: /* @__PURE__ */ t("".concat(o, ",").concat(X), _, r),
  fadeIn100: /* @__PURE__ */ t(c, _, r),
  fadeIn200: /* @__PURE__ */ t(c, l, r),
  fadeIn400: /* @__PURE__ */ t(c, a, r),
  fadeIn500: /* @__PURE__ */ t(c, O, r),
  fadeOut100: /* @__PURE__ */ t(o, _, r),
  fadeOut200: /* @__PURE__ */ t(o, l, r),
  fadeOut400: /* @__PURE__ */ t(o, a, r),
  fadeOut500: /* @__PURE__ */ t(o, O, r),
  rotate90deg: /* @__PURE__ */ t(M, "0.1s", r),
  rotateN90deg: /* @__PURE__ */ t(j, "0.1s", r)
  // expandCollapse 100/200/400, delay 100/200
};
function t(s, E, v) {
  return {
    animationName: s,
    animationDuration: E,
    animationTimingFunction: v,
    animationFillMode: "both"
  };
}
function i(s) {
  return e({
    from: { transform: "translate3d(".concat(s, "px,0,0)"), pointerEvents: "none" },
    to: { transform: "translate3d(0,0,0)", pointerEvents: "auto" }
  });
}
function d(s) {
  return e({
    from: { transform: "translate3d(0,".concat(s, "px,0)"), pointerEvents: "none" },
    to: { transform: "translate3d(0,0,0)", pointerEvents: "auto" }
  });
}
function I(s) {
  return e({
    from: { transform: "translate3d(0,0,0)" },
    to: { transform: "translate3d(".concat(s, "px,0,0)") }
  });
}
function f(s) {
  return e({
    from: { transform: "translate3d(0,0,0)" },
    to: { transform: "translate3d(0,".concat(s, "px,0)") }
  });
}
export {
  B as AnimationStyles
};
