// @fluentui/react/lib/components/Layer/Layer.base.js
import { __assign as K } from "../../../../../tslib/tslib.es6.js";
import * as o from "react";
import * as q from "react-dom";
import { registerLayer as G, unregisterLayer as j, getLayerHost as w, createDefaultLayerHost as z, getDefaultTarget as J } from "./Layer.notification.js";
import { usePortalCompat as Q } from "../../../../react-portal-compat-context/lib/PortalCompatContext.js";
import { FocusRectsContext as X } from "../../../../utilities/lib/useFocusRects.js";
import { IsFocusVisibleClassName as P } from "../../../../utilities/lib/setFocusVisibility.js";
import { useIsomorphicLayoutEffect as Y } from "../../../../utilities/lib/useIsomorphicLayoutEffect.js";
import { FocusRectsProvider as Z } from "../../../../utilities/lib/FocusRectsProvider.js";
import { Fabric as $ } from "../Fabric/Fabric.js";
import { css as ee } from "../../../../utilities/lib/css.js";
import { useWarnings as re } from "../../../../react-hooks/lib/useWarnings.js";
import { classNamesFunction as ne } from "../../../../utilities/lib/classNamesFunction.js";
import { getDocument as te } from "../../../../utilities/lib/dom/getDocument.js";
import { setPortalAttribute as oe } from "../../../../dom-utilities/lib/setPortalAttribute.js";
import { setVirtualParent as ae } from "../../../../dom-utilities/lib/setVirtualParent.js";
import { useMergedRefs as F } from "../../../../react-hooks/lib/useMergedRefs.js";
var ie = /* @__PURE__ */ ne(), ue = function(e) {
  return e != null && e.current ? e.current.classList.contains(P) : !1;
}, le = /* @__PURE__ */ o.forwardRef(function(e, B) {
  var I = Q(), l = o.useRef(null), x = F(l, B), u = o.useRef(), O = o.useRef(null), v = o.useContext(X), D = o.useState(!1), p = D[0], N = D[1], S = o.useCallback(function(r) {
    var n = ue(v == null ? void 0 : v.providerRef);
    r && n && r.classList.add(P);
  }, [v]), T = e.children, U = e.className, V = e.eventBubblingEnabled, s = e.fabricProps, i = e.hostId, H = e.insertFirst, M = e.onLayerDidMount, g = M === void 0 ? function() {
  } : M, _ = e.onLayerMounted, y = _ === void 0 ? function() {
  } : _, b = e.onLayerWillUnmount, W = e.styles, k = e.theme, R = F(O, s == null ? void 0 : s.ref, S), h = ie(W, {
    theme: k,
    className: U,
    isNotHost: !i
  }), A = function(r, n) {
    var c, f;
    n === void 0 && (n = null);
    var d = n ?? r;
    if (i) {
      var m = w(i);
      return m ? (c = m.rootRef.current) !== null && c !== void 0 ? c : null : (f = d.getElementById(i)) !== null && f !== void 0 ? f : null;
    } else {
      var t = J(), a = t ? d.querySelector(t) : null;
      return a || (a = z(r, n)), a;
    }
  }, C = function() {
    b == null || b();
    var r = u.current;
    u.current = void 0, r && r.parentNode && r.parentNode.removeChild(r);
  }, L = function() {
    var r, n, c, f, d = te(l.current), m = !((n = (r = l.current) === null || r === void 0 ? void 0 : r.getRootNode()) === null || n === void 0) && n.host ? (c = l == null ? void 0 : l.current) === null || c === void 0 ? void 0 : c.getRootNode() : void 0;
    if (!(!d || !d && !m)) {
      var t = A(d, m);
      if (t) {
        t.__tabsterElementFlags || (t.__tabsterElementFlags = {}), t.__tabsterElementFlags.noDirectAriaHidden = !0, C();
        var a = ((f = t.ownerDocument) !== null && f !== void 0 ? f : d).createElement("div");
        a.className = h.root, oe(a), ae(a, l.current), H ? t.insertBefore(a, t.firstChild) : t.appendChild(a), u.current = a, N(!0);
      }
    }
  };
  return Y(function() {
    L(), i && G(i, L);
    var r = u.current ? I(u.current) : void 0;
    return function() {
      r && r(), C(), i && j(i, L);
    };
  }, [i]), o.useEffect(function() {
    u.current && p && (y == null || y(), g == null || g(), N(!1));
  }, [p, y, g]), fe(e), o.createElement("span", { className: "ms-layer", ref: x }, u.current && q.createPortal(o.createElement(
    Z,
    { layerRoot: !0, providerRef: R },
    o.createElement($, K({}, !V && ce(), s, { className: ee(h.content, s == null ? void 0 : s.className), ref: R }), T)
  ), u.current));
});
le.displayName = "LayerBase";
var E, se = function(e) {
  e.eventPhase === Event.BUBBLING_PHASE && e.type !== "mouseenter" && e.type !== "mouseleave" && e.type !== "touchstart" && e.type !== "touchend" && e.stopPropagation();
};
function ce() {
  return E || (E = {}, [
    "onClick",
    "onContextMenu",
    "onDoubleClick",
    "onDrag",
    "onDragEnd",
    "onDragEnter",
    "onDragExit",
    "onDragLeave",
    "onDragOver",
    "onDragStart",
    "onDrop",
    "onMouseDown",
    "onMouseEnter",
    "onMouseLeave",
    "onMouseMove",
    "onMouseOver",
    "onMouseOut",
    "onMouseUp",
    "onTouchMove",
    "onTouchStart",
    "onTouchCancel",
    "onTouchEnd",
    "onKeyDown",
    "onKeyPress",
    "onKeyUp",
    "onFocus",
    "onBlur",
    "onChange",
    "onInput",
    "onInvalid",
    "onSubmit"
  ].forEach(function(e) {
    return E[e] = se;
  })), E;
}
function fe(e) {
  process.env.NODE_ENV !== "production" && re({
    name: "Layer",
    props: e,
    deprecations: { onLayerMounted: "onLayerDidMount" }
  });
}
export {
  le as LayerBase
};
