// @fluentui/react/lib/components/Layer/Layer.notification.js
var f = {}, l = {}, v = "fluent-default-layer-host", y = /* @__PURE__ */ "#".concat(v);
function d(e, t) {
  f[e] || (f[e] = []), f[e].push(t);
  var r = l[e];
  if (r)
    for (var a = 0, n = r; a < n.length; a++) {
      var i = n[a];
      i.notifyLayersChanged();
    }
}
function g(e, t) {
  var r = f[e];
  if (r) {
    var a = r.indexOf(t);
    a >= 0 && (r.splice(a, 1), r.length === 0 && delete f[e]);
  }
  var n = l[e];
  if (n)
    for (var i = 0, u = n; i < u.length; i++) {
      var s = u[i];
      s.notifyLayersChanged();
    }
}
function o(e) {
  var t = l[e];
  return t && t[0] || void 0;
}
function c(e, t) {
  t === void 0 && (t = null);
  var r = e.createElement("div");
  return r.setAttribute("id", v), r.style.cssText = "position:fixed;z-index:1000000", t ? t.appendChild(r) : e == null || e.body.appendChild(r), r;
}
function p() {
  return y;
}
export {
  c as createDefaultLayerHost,
  p as getDefaultTarget,
  o as getLayerHost,
  d as registerLayer,
  g as unregisterLayer
};
