// axios/lib/defaults/index.js
import t from "../utils.js";
import p from "../core/AxiosError.js";
import a from "./transitional.js";
import u from "../helpers/toFormData.js";
import S from "../helpers/toURLEncodedForm.js";
import l from "../platform/index.js";
import h from "../helpers/formDataToJSON.js";
function O(n, r, e) {
  if (t.isString(n))
    try {
      return (r || JSON.parse)(n), t.trim(n);
    } catch (i) {
      if (i.name !== "SyntaxError")
        throw i;
    }
  return (e || JSON.stringify)(n);
}
const y = {
  transitional: a,
  adapter: ["xhr", "http", "fetch"],
  transformRequest: [function(r, e) {
    const i = e.getContentType() || "", s = i.indexOf("application/json") > -1, f = t.isObject(r);
    if (f && t.isHTMLForm(r) && (r = new FormData(r)), t.isFormData(r))
      return s ? JSON.stringify(h(r)) : r;
    if (t.isArrayBuffer(r) || t.isBuffer(r) || t.isStream(r) || t.isFile(r) || t.isBlob(r) || t.isReadableStream(r))
      return r;
    if (t.isArrayBufferView(r))
      return r.buffer;
    if (t.isURLSearchParams(r))
      return e.setContentType("application/x-www-form-urlencoded;charset=utf-8", !1), r.toString();
    let o;
    if (f) {
      if (i.indexOf("application/x-www-form-urlencoded") > -1)
        return S(r, this.formSerializer).toString();
      if ((o = t.isFileList(r)) || i.indexOf("multipart/form-data") > -1) {
        const c = this.env && this.env.FormData;
        return u(
          o ? { "files[]": r } : r,
          c && new c(),
          this.formSerializer
        );
      }
    }
    return f || s ? (e.setContentType("application/json", !1), O(r)) : r;
  }],
  transformResponse: [function(r) {
    const e = this.transitional || y.transitional, i = e && e.forcedJSONParsing, s = this.responseType === "json";
    if (t.isResponse(r) || t.isReadableStream(r))
      return r;
    if (r && t.isString(r) && (i && !this.responseType || s)) {
      const m = !(e && e.silentJSONParsing) && s;
      try {
        return JSON.parse(r);
      } catch (o) {
        if (m)
          throw o.name === "SyntaxError" ? p.from(o, p.ERR_BAD_RESPONSE, this, null, this.response) : o;
      }
    }
    return r;
  }],
  /**
   * A timeout in milliseconds to abort a request. If set to 0 (default) a
   * timeout is not created.
   */
  timeout: 0,
  xsrfCookieName: "XSRF-TOKEN",
  xsrfHeaderName: "X-XSRF-TOKEN",
  maxContentLength: -1,
  maxBodyLength: -1,
  env: {
    FormData: l.classes.FormData,
    Blob: l.classes.Blob
  },
  validateStatus: function(r) {
    return r >= 200 && r < 300;
  },
  headers: {
    common: {
      Accept: "application/json, text/plain, */*",
      "Content-Type": void 0
    }
  }
};
export {
  y as default
};
