// @fluentui/utilities/lib/Async.js
import { getWindow as w } from "./dom/getWindow.js";
var k = (
  /** @class */
  /* @__PURE__ */ function() {
    function n(t, r) {
      this._timeoutIds = null, this._immediateIds = null, this._intervalIds = null, this._animationFrameIds = null, this._isDisposed = !1, this._parent = t || null, this._onErrorHandler = r, this._noop = function() {
      };
    }
    return n.prototype.dispose = function() {
      var t;
      if (this._isDisposed = !0, this._parent = null, this._timeoutIds) {
        for (t in this._timeoutIds)
          this._timeoutIds.hasOwnProperty(t) && this.clearTimeout(parseInt(t, 10));
        this._timeoutIds = null;
      }
      if (this._immediateIds) {
        for (t in this._immediateIds)
          this._immediateIds.hasOwnProperty(t) && this.clearImmediate(parseInt(t, 10));
        this._immediateIds = null;
      }
      if (this._intervalIds) {
        for (t in this._intervalIds)
          this._intervalIds.hasOwnProperty(t) && this.clearInterval(parseInt(t, 10));
        this._intervalIds = null;
      }
      if (this._animationFrameIds) {
        for (t in this._animationFrameIds)
          this._animationFrameIds.hasOwnProperty(t) && this.cancelAnimationFrame(parseInt(t, 10));
        this._animationFrameIds = null;
      }
    }, n.prototype.setTimeout = function(t, r) {
      var i = this, e = 0;
      return this._isDisposed || (this._timeoutIds || (this._timeoutIds = {}), e = setTimeout(function() {
        try {
          i._timeoutIds && delete i._timeoutIds[e], t.apply(i._parent);
        } catch (a) {
          i._logError(a);
        }
      }, r), this._timeoutIds[e] = !0), e;
    }, n.prototype.clearTimeout = function(t) {
      this._timeoutIds && this._timeoutIds[t] && (clearTimeout(t), delete this._timeoutIds[t]);
    }, n.prototype.setImmediate = function(t, r) {
      var i = this, e = 0, a = w(r);
      if (!this._isDisposed) {
        this._immediateIds || (this._immediateIds = {});
        var s = function() {
          try {
            i._immediateIds && delete i._immediateIds[e], t.apply(i._parent);
          } catch (m) {
            i._logError(m);
          }
        };
        e = a.setTimeout(s, 0), this._immediateIds[e] = !0;
      }
      return e;
    }, n.prototype.clearImmediate = function(t, r) {
      var i = w(r);
      this._immediateIds && this._immediateIds[t] && (i.clearTimeout(t), delete this._immediateIds[t]);
    }, n.prototype.setInterval = function(t, r) {
      var i = this, e = 0;
      return this._isDisposed || (this._intervalIds || (this._intervalIds = {}), e = setInterval(function() {
        try {
          t.apply(i._parent);
        } catch (a) {
          i._logError(a);
        }
      }, r), this._intervalIds[e] = !0), e;
    }, n.prototype.clearInterval = function(t) {
      this._intervalIds && this._intervalIds[t] && (clearInterval(t), delete this._intervalIds[t]);
    }, n.prototype.throttle = function(t, r, i) {
      var e = this;
      if (this._isDisposed)
        return this._noop;
      var a = r || 0, s = !0, m = !0, v = 0, h, c, f = null;
      i && typeof i.leading == "boolean" && (s = i.leading), i && typeof i.trailing == "boolean" && (m = i.trailing);
      var _ = function(l) {
        var o = Date.now(), I = o - v, y = s ? a - I : a;
        return I >= a && (!l || s) ? (v = o, f && (e.clearTimeout(f), f = null), h = t.apply(e._parent, c)) : f === null && m && (f = e.setTimeout(_, y)), h;
      }, p = function() {
        for (var l = [], o = 0; o < arguments.length; o++)
          l[o] = arguments[o];
        return c = l, _(!0);
      };
      return p;
    }, n.prototype.debounce = function(t, r, i) {
      var e = this;
      if (this._isDisposed) {
        var a = function() {
        };
        return a.cancel = function() {
        }, a.flush = function() {
          return null;
        }, a.pending = function() {
          return !1;
        }, a;
      }
      var s = r || 0, m = !1, v = !0, h = null, c = 0, f = Date.now(), _, p, l = null;
      i && typeof i.leading == "boolean" && (m = i.leading), i && typeof i.trailing == "boolean" && (v = i.trailing), i && typeof i.maxWait == "number" && !isNaN(i.maxWait) && (h = i.maxWait);
      var o = function(d) {
        l && (e.clearTimeout(l), l = null), f = d;
      }, I = function(d) {
        o(d), _ = t.apply(e._parent, p);
      }, y = function(d) {
        var u = Date.now(), D = !1;
        d && (m && u - c >= s && (D = !0), c = u);
        var x = u - c, T = s - x, E = u - f, A = !1;
        return h !== null && (E >= h && l ? A = !0 : T = Math.min(T, h - E)), x >= s || A || D ? I(u) : (l === null || !d) && v && (l = e.setTimeout(y, T)), _;
      }, F = function() {
        return !!l;
      }, b = function() {
        F() && o(Date.now());
      }, W = function() {
        return F() && I(Date.now()), _;
      }, g = function() {
        for (var d = [], u = 0; u < arguments.length; u++)
          d[u] = arguments[u];
        return p = d, y(!0);
      };
      return g.cancel = b, g.flush = W, g.pending = F, g;
    }, n.prototype.requestAnimationFrame = function(t, r) {
      var i = this, e = 0, a = w(r);
      if (!this._isDisposed) {
        this._animationFrameIds || (this._animationFrameIds = {});
        var s = function() {
          try {
            i._animationFrameIds && delete i._animationFrameIds[e], t.apply(i._parent);
          } catch (m) {
            i._logError(m);
          }
        };
        e = a.requestAnimationFrame ? a.requestAnimationFrame(s) : a.setTimeout(s, 0), this._animationFrameIds[e] = !0;
      }
      return e;
    }, n.prototype.cancelAnimationFrame = function(t, r) {
      var i = w(r);
      this._animationFrameIds && this._animationFrameIds[t] && (i.cancelAnimationFrame ? i.cancelAnimationFrame(t) : i.clearTimeout(t), delete this._animationFrameIds[t]);
    }, n.prototype._logError = function(t) {
      this._onErrorHandler && this._onErrorHandler(t);
    }, n;
  }()
);
export {
  k as Async
};
