// @fluentui/utilities/lib/classNamesFunction.js
import { getRTL as M } from "./rtl.js";
import { Stylesheet as R } from "../../merge-styles/lib/Stylesheet.js";
import { mergeCssSets as S } from "../../merge-styles/lib/mergeStyleSets.js";
import { getWindow as I } from "./dom/getWindow.js";
var N = 50, E = 5, h = 0;
var o = "__retval__";
function V(e) {
  e === void 0 && (e = {});
  var a = /* @__PURE__ */ new Map(), r = 0, _ = 0, t = h, l = function(f, d) {
    var s;
    if (d === void 0 && (d = {}), e.useStaticStyles && typeof f == "function" && f.__noStyleOverride__)
      return f(d);
    _++;
    var v = f ? f.__shadowConfig__ : void 0, n = v && v.window ? v.window : "__default__";
    a.has(n) || a.set(n, /* @__PURE__ */ new Map());
    var i = a.get(n), C = d.theme, u = C && C.rtl !== void 0 ? C.rtl : M(), c = e.disableCaching;
    if (t !== h && (t = h, a.set(n, /* @__PURE__ */ new Map()), i = a.get(n), r = 0), e.disableCaching || (i = w(a.get(n), f), i = w(i, d)), (c || !i[o]) && (f === void 0 ? i[o] = {} : i[o] = S([
      typeof f == "function" ? f(d) : f
    ], {
      shadowConfig: f.__shadowConfig__,
      rtl: !!u,
      specificityMultiplier: e.useStaticStyles ? E : void 0
    }), c || r++), r > (e.cacheSize || N)) {
      var g = I();
      !((s = g == null ? void 0 : g.FabricConfig) === null || s === void 0) && s.enableClassNameCacheFullWarning && (console.warn("Styles are being recalculated too frequently. Cache miss rate is ".concat(r, "/").concat(_, ".")), console.trace()), a.get(n).clear(), r = 0, e.disableCaching = !0;
    }
    return i[o];
  };
  return l;
}
function m(e, a) {
  return a = T(a), e.has(a) || e.set(a, /* @__PURE__ */ new Map()), e.get(a);
}
function w(e, a) {
  if (typeof a == "function") {
    var r = a.__cachedInputs__;
    if (r)
      for (var _ = 0, t = a.__cachedInputs__; _ < t.length; _++) {
        var l = t[_];
        e = m(e, l);
      }
    else
      e = m(e, a);
  } else if (typeof a == "object")
    for (var f in a)
      a.hasOwnProperty(f) && (e = m(e, a[f]));
  return e;
}
function T(e) {
  switch (e) {
    case void 0:
      return "__undefined__";
    case null:
      return "__null__";
    default:
      return e;
  }
}
export {
  V as classNamesFunction
};
