// @fluentui/utilities/lib/EventGroup.js
import { assign as d } from "./object.js";
import { getDocument as b } from "./dom/getDocument.js";
var C = (
  /** @class */
  /* @__PURE__ */ function() {
    function f(i) {
      this._id = f._uniqueId++, this._parent = i, this._eventRecords = [];
    }
    return f.raise = function(i, t, s, e, u) {
      var n, a = u ?? b();
      if (f._isElement(i)) {
        if (typeof a < "u" && a.createEvent) {
          var l = a.createEvent("HTMLEvents");
          l.initEvent(t, e || !1, !0), d(l, s), n = i.dispatchEvent(l);
        } else if (typeof a < "u" && a.createEventObject) {
          var o = a.createEventObject(s);
          i.fireEvent("on" + t, o);
        }
      } else
        for (; i && n !== !1; ) {
          var h = i.__events__, c = h ? h[t] : null;
          if (c) {
            for (var E in c)
              if (c.hasOwnProperty(E))
                for (var p = c[E], r = 0; n !== !1 && r < p.length; r++) {
                  var _ = p[r];
                  _.objectCallback && (n = _.objectCallback.call(_.parent, s));
                }
          }
          i = e ? i.parent : null;
        }
      return n;
    }, f.isObserved = function(i, t) {
      var s = i && i.__events__;
      return !!s && !!s[t];
    }, f.isDeclared = function(i, t) {
      var s = i && i.__declaredEvents;
      return !!s && !!s[t];
    }, f.stopPropagation = function(i) {
      i.stopPropagation ? i.stopPropagation() : i.cancelBubble = !0;
    }, f._isElement = function(i) {
      return !!i && (!!i.addEventListener || typeof HTMLElement < "u" && i instanceof HTMLElement);
    }, f.prototype.dispose = function() {
      this._isDisposed || (this._isDisposed = !0, this.off(), this._parent = null);
    }, f.prototype.onAll = function(i, t, s) {
      for (var e in t)
        t.hasOwnProperty(e) && this.on(i, e, t[e], s);
    }, f.prototype.on = function(i, t, s, e) {
      var u = this;
      if (t.indexOf(",") > -1)
        for (var n = t.split(/[ ,]+/), a = 0; a < n.length; a++)
          this.on(i, n[a], s, e);
      else {
        var l = this._parent, o = {
          target: i,
          eventName: t,
          parent: l,
          callback: s,
          options: e
        }, n = i.__events__ = i.__events__ || {};
        if (n[t] = n[t] || {
          count: 0
        }, n[t][this._id] = n[t][this._id] || [], n[t][this._id].push(o), n[t].count++, f._isElement(i)) {
          var h = function() {
            for (var p = [], r = 0; r < arguments.length; r++)
              p[r] = arguments[r];
            if (!u._isDisposed) {
              var _;
              try {
                if (_ = s.apply(l, p), _ === !1 && p[0]) {
                  var v = p[0];
                  v.preventDefault && v.preventDefault(), v.stopPropagation && v.stopPropagation(), v.cancelBubble = !0;
                }
              } catch {
              }
              return _;
            }
          };
          o.elementCallback = h, i.addEventListener ? i.addEventListener(t, h, e) : i.attachEvent && i.attachEvent("on" + t, h);
        } else {
          var c = function() {
            for (var p = [], r = 0; r < arguments.length; r++)
              p[r] = arguments[r];
            if (!u._isDisposed)
              return s.apply(l, p);
          };
          o.objectCallback = c;
        }
        this._eventRecords.push(o);
      }
    }, f.prototype.off = function(i, t, s, e) {
      for (var u = 0; u < this._eventRecords.length; u++) {
        var n = this._eventRecords[u];
        if ((!i || i === n.target) && (!t || t === n.eventName) && (!s || s === n.callback) && (typeof e != "boolean" || e === n.options)) {
          var a = n.target.__events__, l = a[n.eventName], o = l ? l[this._id] : null;
          o && (o.length === 1 || !s ? (l.count -= o.length, delete a[n.eventName][this._id]) : (l.count--, o.splice(o.indexOf(n), 1)), l.count || delete a[n.eventName]), n.elementCallback && (n.target.removeEventListener ? n.target.removeEventListener(n.eventName, n.elementCallback, n.options) : n.target.detachEvent && n.target.detachEvent("on" + n.eventName, n.elementCallback)), this._eventRecords.splice(u--, 1);
        }
      }
    }, f.prototype.raise = function(i, t, s) {
      return f.raise(this._parent, i, t, s);
    }, f.prototype.declare = function(i) {
      var t = this._parent.__declaredEvents = this._parent.__declaredEvents || {};
      if (typeof i == "string")
        t[i] = !0;
      else
        for (var s = 0; s < i.length; s++)
          t[i[s]] = !0;
    }, f._uniqueId = 0, f;
  }()
);
export {
  C as EventGroup
};
