// @fluentui/utilities/lib/focus.js
import { getWindow as B } from "./dom/getWindow.js";
import { getDocument as V } from "./dom/getDocument.js";
import { elementContainsAttribute as F } from "../../dom-utilities/lib/elementContainsAttribute.js";
import { getParent as P } from "../../dom-utilities/lib/getParent.js";
import { elementContains as p } from "../../dom-utilities/lib/elementContains.js";
var y = "data-is-focusable", L = "data-is-visible", D = "data-focuszone-id", S = "data-is-sub-focuszone";
function c(i, r, t, u) {
  return g(i, r, !0, !1, !1, t, void 0, void 0, void 0, u);
}
function j(i, r, t, u) {
  return A(i, r, !0, !1, !0, t, void 0, void 0, u);
}
function G(i, r, t) {
  var u = g(i, i, !0, !1, !1, !0, void 0, void 0, r, t);
  return u ? (_(u), !0) : !1;
}
function A(i, r, t, u, n, f, a, s, v) {
  if (!r || r === i)
    return null;
  var e = I(r);
  if (n && e && (f || !(h(r) || O(r)))) {
    var C = r.lastElementChild || v, d = A(i, C, !0, !0, !0, f, a, s, v);
    if (d)
      return d;
  }
  if (t && e && N(r, s, v))
    return r;
  var b = A(i, r.previousElementSibling, !0, !0, !0, f, a, s, v);
  return b || (u ? null : A(i, r.parentElement, !0, !1, !1, f, a, s, v));
}
function g(i, r, t, u, n, f, a, s, v, e) {
  if (!r || r === i && n && !a)
    return null;
  var C = v ? W : I, d = C(r);
  if (t && d && N(r, s, e))
    return r;
  if (!n && d && (f || !(h(r) || O(r)))) {
    var b = r.firstElementChild || e, x = g(i, b, !0, !0, !1, f, a, s, v, e);
    if (x)
      return x;
  }
  if (r === i)
    return null;
  var U = g(i, r.nextElementSibling, !0, !0, !1, f, a, s, v, e);
  return U || (u ? null : g(i, r.parentElement, !1, !1, !0, f, a, s, v, e));
}
function I(i) {
  if (!i || !i.getAttribute)
    return !1;
  var r = i.getAttribute(L);
  return r != null ? r === "true" : i.offsetHeight !== 0 || i.offsetParent !== null || // eslint-disable-next-line @typescript-eslint/no-explicit-any
  i.isVisible === !0;
}
function W(i, r) {
  var t = r ?? B();
  return !!i && I(i) && !i.hidden && t.getComputedStyle(i).visibility !== "hidden";
}
function N(i, r, t) {
  if (t === void 0 && (t = !0), !i || i.disabled)
    return !1;
  var u = 0, n = null;
  i && i.getAttribute && (n = i.getAttribute("tabIndex"), n && (u = parseInt(n, 10)));
  var f = i.getAttribute ? i.getAttribute(y) : null, a = n !== null && u >= 0, s = t && i.shadowRoot ? !!i.shadowRoot.delegatesFocus : !1, v = !!i && f !== "false" && (i.tagName === "A" || i.tagName === "BUTTON" || i.tagName === "INPUT" || i.tagName === "TEXTAREA" || i.tagName === "SELECT" || f === "true" || a || s);
  return v;
}
function h(i) {
  return !!(i && i.getAttribute && i.getAttribute(D));
}
function O(i) {
  return !!(i && i.getAttribute && i.getAttribute(S) === "true");
}
function J(i) {
  var r = V(i), t = r && r.activeElement;
  return !!(t && p(i, t));
}
function K(i, r, t) {
  var u = V();
  return F(i, r, u) !== "true";
}
var T = void 0;
function _(i) {
  if (i) {
    var r = B(i);
    r && (T !== void 0 && r.cancelAnimationFrame(T), T = r.requestAnimationFrame(function() {
      i && i.focus(), T = void 0;
    }));
  }
}
function Q(i, r) {
  for (var t = i, u = 0, n = r; u < n.length; u++) {
    var f = n[u], a = t.children[Math.min(f, t.children.length - 1)];
    if (!a)
      break;
    t = a;
  }
  return t = N(t) && I(t) ? t : g(i, t, !0) || A(i, t), t;
}
function Y(i, r) {
  for (var t = []; r && i && r !== i; ) {
    var u = P(r, !0);
    if (u === null)
      return [];
    t.unshift(Array.prototype.indexOf.call(u.children, r)), r = u;
  }
  return t;
}
export {
  J as doesElementContainFocus,
  _ as focusAsync,
  G as focusFirstChild,
  Y as getElementIndexPath,
  c as getFirstFocusable,
  Q as getFocusableByIndexPath,
  j as getLastFocusable,
  g as getNextElement,
  A as getPreviousElement,
  O as isElementFocusSubZone,
  h as isElementFocusZone,
  N as isElementTabbable,
  I as isElementVisible,
  W as isElementVisibleAndNotHidden,
  K as shouldWrapFocus
};
