// @fluentui/utilities/lib/properties.js
var r = function() {
  for (var i = [], o = 0; o < arguments.length; o++)
    i[o] = arguments[o];
  for (var n = {}, a = 0, l = i; a < l.length; a++)
    for (var s = l[a], u = Array.isArray(s) ? s : Object.keys(s), t = 0, e = u; t < e.length; t++) {
      var d = e[t];
      n[d] = 1;
    }
  return n;
}, c = /* @__PURE__ */ r([
  "onCopy",
  "onCut",
  "onPaste",
  "onCompositionEnd",
  "onCompositionStart",
  "onCompositionUpdate",
  "onFocus",
  "onFocusCapture",
  "onBlur",
  "onBlurCapture",
  "onChange",
  "onInput",
  "onSubmit",
  "onLoad",
  "onError",
  "onKeyDown",
  "onKeyDownCapture",
  "onKeyPress",
  "onKeyUp",
  "onAbort",
  "onCanPlay",
  "onCanPlayThrough",
  "onDurationChange",
  "onEmptied",
  "onEncrypted",
  "onEnded",
  "onLoadedData",
  "onLoadedMetadata",
  "onLoadStart",
  "onPause",
  "onPlay",
  "onPlaying",
  "onProgress",
  "onRateChange",
  "onSeeked",
  "onSeeking",
  "onStalled",
  "onSuspend",
  "onTimeUpdate",
  "onVolumeChange",
  "onWaiting",
  "onClick",
  "onClickCapture",
  "onContextMenu",
  "onDoubleClick",
  "onDrag",
  "onDragEnd",
  "onDragEnter",
  "onDragExit",
  "onDragLeave",
  "onDragOver",
  "onDragStart",
  "onDrop",
  "onMouseDown",
  "onMouseDownCapture",
  "onMouseEnter",
  "onMouseLeave",
  "onMouseMove",
  "onMouseOut",
  "onMouseOver",
  "onMouseUp",
  "onMouseUpCapture",
  "onSelect",
  "onTouchCancel",
  "onTouchEnd",
  "onTouchMove",
  "onTouchStart",
  "onScroll",
  "onWheel",
  "onPointerCancel",
  "onPointerDown",
  "onPointerEnter",
  "onPointerLeave",
  "onPointerMove",
  "onPointerOut",
  "onPointerOver",
  "onPointerUp",
  "onGotPointerCapture",
  "onLostPointerCapture"
]), h = /* @__PURE__ */ r([
  "accessKey",
  "children",
  "className",
  "contentEditable",
  "dir",
  "draggable",
  "hidden",
  "htmlFor",
  "id",
  "lang",
  "ref",
  "role",
  "style",
  "tabIndex",
  "title",
  "translate",
  "spellCheck",
  "name"
  // global
]), p = /* @__PURE__ */ r(h, c), g = /* @__PURE__ */ r(p, [
  "height",
  "loop",
  "muted",
  "preload",
  "src",
  "width"
  // canvas, embed, iframe, img, input, object, video
]), v = /* @__PURE__ */ r(p, [
  "download",
  "href",
  "hrefLang",
  "media",
  "rel",
  "target",
  "type"
  // a, button, input, link, menu, object, script, source, style
]), m = /* @__PURE__ */ r(p, [
  "autoFocus",
  "disabled",
  "form",
  "formAction",
  "formEncType",
  "formMethod",
  "formNoValidate",
  "formTarget",
  "type",
  "value"
  // button, input, li, option, meter, progress, param,
]), C = /* @__PURE__ */ r(p, [
  "alt",
  "crossOrigin",
  "height",
  "src",
  "srcSet",
  "useMap",
  "width"
  // canvas, embed, iframe, img, input, object, video
]), f = p;
function P(i, o, n) {
  for (var a = Array.isArray(o), l = {}, s = Object.keys(i), u = 0, t = s; u < t.length; u++) {
    var e = t[u], d = !a && o[e] || a && o.indexOf(e) >= 0 || e.indexOf("data-") === 0 || e.indexOf("aria-") === 0;
    d && (!n || (n == null ? void 0 : n.indexOf(e)) === -1) && (l[e] = i[e]);
  }
  return l;
}
export {
  v as anchorProperties,
  g as audioProperties,
  c as baseElementEvents,
  h as baseElementProperties,
  m as buttonProperties,
  f as divProperties,
  P as getNativeProps,
  p as htmlElementProperties,
  C as imgProperties
};
