// @fluentui/utilities/lib/Rectangle.js
var a = (
  /** @class */
  /* @__PURE__ */ function() {
    function e(t, o, i, r) {
      t === void 0 && (t = 0), o === void 0 && (o = 0), i === void 0 && (i = 0), r === void 0 && (r = 0), this.top = i, this.bottom = r, this.left = t, this.right = o;
    }
    return Object.defineProperty(e.prototype, "width", {
      /**
       * Calculated automatically by subtracting the right from left
       */
      get: function() {
        return this.right - this.left;
      },
      enumerable: !1,
      configurable: !0
    }), Object.defineProperty(e.prototype, "height", {
      /**
       * Calculated automatically by subtracting the bottom from top.
       */
      get: function() {
        return this.bottom - this.top;
      },
      enumerable: !1,
      configurable: !0
    }), e.prototype.equals = function(t) {
      return parseFloat(this.top.toFixed(4)) === parseFloat(t.top.toFixed(4)) && parseFloat(this.bottom.toFixed(4)) === parseFloat(t.bottom.toFixed(4)) && parseFloat(this.left.toFixed(4)) === parseFloat(t.left.toFixed(4)) && parseFloat(this.right.toFixed(4)) === parseFloat(t.right.toFixed(4));
    }, e;
  }()
);
export {
  a as Rectangle
};
