// @fluentui/react/lib/components/Popup/Popup.js
import { __assign as p } from "../../../../../tslib/tslib.es6.js";
import * as t from "react";
import { getPropsWithDefaults as R } from "../../../../utilities/lib/getPropsWithDefaults.js";
import { useMergedRefs as w } from "../../../../react-hooks/lib/useMergedRefs.js";
import { KeyCodes as S } from "../../../../utilities/lib/KeyCodes.js";
import { useWindow as h } from "../../../../react-window-provider/lib/WindowProvider.js";
import { useOnEvent as m } from "../../../../react-hooks/lib/useOnEvent.js";
import { getNativeProps as D, divProperties as B } from "../../../../utilities/lib/properties.js";
import { useAsync as C } from "../../../../react-hooks/lib/useAsync.js";
import { getDocument as b } from "../../../../utilities/lib/dom/getDocument.js";
import { doesElementContainFocus as N } from "../../../../utilities/lib/focus.js";
import { modalize as _ } from "../../../../utilities/lib/modalize.js";
import { getWindow as P } from "../../../../utilities/lib/dom/getWindow.js";
function k(a, r) {
  var e = C(), n = t.useState(!1), o = n[0], s = n[1];
  return t.useEffect(function() {
    return e.requestAnimationFrame(function() {
      var i;
      if (!(a.style && a.style.overflowY)) {
        var l = !1;
        if (r && r.current && (!((i = r.current) === null || i === void 0) && i.firstElementChild)) {
          var u = r.current.clientHeight, c = r.current.firstElementChild.clientHeight;
          u > 0 && c > u && (l = c - u > 1);
        }
        o !== l && s(l);
      }
    }), function() {
      return e.dispose();
    };
  }), o;
}
function A(a) {
  var r = a.originalElement, e = a.containsFocus;
  r && e && r !== P() && setTimeout(function() {
    var n;
    (n = r.focus) === null || n === void 0 || n.call(r);
  }, 0);
}
function L(a, r) {
  var e = a.onRestoreFocus, n = e === void 0 ? A : e, o = t.useRef(), s = t.useRef(!1);
  t.useEffect(function() {
    return o.current = b().activeElement, N(r.current) && (s.current = !0), function() {
      var i;
      n == null || n({
        originalElement: o.current,
        containsFocus: s.current,
        documentContainsFocus: ((i = b()) === null || i === void 0 ? void 0 : i.hasFocus()) || !1
      }), o.current = void 0;
    };
  }, []), m(r, "focus", t.useCallback(function() {
    s.current = !0;
  }, []), !0), m(r, "blur", t.useCallback(function(i) {
    r.current && i.relatedTarget && !r.current.contains(i.relatedTarget) && (s.current = !1);
  }, []), !0);
}
function V(a, r) {
  var e = String(a["aria-modal"]).toLowerCase() === "true" && a.enableAriaHiddenSiblings;
  t.useEffect(function() {
    if (e && r.current) {
      var n = _(r.current);
      return n;
    }
  }, [r, e]);
}
var H = /* @__PURE__ */ t.forwardRef(function(a, r) {
  var e = R({ shouldRestoreFocus: !0, enableAriaHiddenSiblings: !0 }, a), n = t.useRef(), o = w(n, r);
  V(e, n), L(e, n);
  var s = e.role, i = e.className, l = e.ariaLabel, u = e.ariaLabelledBy, c = e.ariaDescribedBy, y = e.style, g = e.children, d = e.onDismiss, F = k(e, n), v = t.useCallback(function(f) {
    switch (f.which) {
      case S.escape:
        d && (d(f), f.preventDefault(), f.stopPropagation());
        break;
    }
  }, [d]), E = h();
  return m(E, "keydown", v), t.createElement("div", p({ ref: o }, D(e, B), { className: i, role: s, "aria-label": l, "aria-labelledby": u, "aria-describedby": c, onKeyDown: v, style: p({ overflowY: F ? "scroll" : void 0, outline: "none" }, y) }), g);
});
H.displayName = "Popup";
export {
  H as Popup
};
