// @fluentui/react/lib/components/Fabric/Fabric.base.js
import { __assign as f } from "../../../../../tslib/tslib.es6.js";
import * as o from "react";
import { classNamesFunction as y } from "../../../../utilities/lib/classNamesFunction.js";
import { memoizeFunction as g } from "../../../../utilities/lib/memoize.js";
import { getNativeProps as p, divProperties as N } from "../../../../utilities/lib/properties.js";
import { FocusRectsProvider as R } from "../../../../utilities/lib/FocusRectsProvider.js";
import { Customizer as b } from "../../../../utilities/lib/customizations/Customizer.js";
import { getDocument as F } from "../../../../utilities/lib/dom/getDocument.js";
import { getRTL as v } from "../../../../utilities/lib/rtl.js";
import { createTheme as D } from "../../../../theme/lib/createTheme.js";
import { useMergedRefs as B } from "../../../../react-hooks/lib/useMergedRefs.js";
var C = /* @__PURE__ */ y(), P = /* @__PURE__ */ g(function(e, m) {
  return D(f(f({}, e), { rtl: m }));
}), E = function(e) {
  var m = e.theme, r = e.dir, t = v(m) ? "rtl" : "ltr", a = v() ? "rtl" : "ltr", i = r || t;
  return {
    // If Fabric dir !== contextDir
    // Or If contextDir !== pageDir
    // Then we need to set dir of the Fabric root
    rootDir: i !== t || i !== a ? i : r,
    // If dir !== contextDir || pageDir
    // then set contextual theme around content
    needsTheme: i !== t
  };
}, L = /* @__PURE__ */ o.forwardRef(function(e, m) {
  var r = e.className, t = e.theme, a = e.applyTheme, i = e.applyThemeToBody, d = e.styles, n = C(d, {
    theme: t,
    applyTheme: a,
    className: r
  }), s = o.useRef(null);
  return z(i, n, s), o.createElement(o.Fragment, null, x(e, n, s, m));
});
L.displayName = "FabricBase";
function x(e, m, r, t) {
  var a = m.root, i = e.as, d = i === void 0 ? "div" : i, n = e.dir, s = e.theme, u = p(e, N, ["dir"]), l = E(e), h = l.rootDir, T = l.needsTheme, c = o.createElement(
    R,
    { providerRef: r },
    o.createElement(d, f({ dir: h }, u, { className: a, ref: B(r, t) }))
  );
  return T && (c = // eslint-disable-next-line deprecation/deprecation
  o.createElement(b, { settings: { theme: P(s, n === "rtl") } }, c)), c;
}
function z(e, m, r) {
  var t = m.bodyThemed;
  return o.useEffect(function() {
    if (e) {
      var a = F(r.current);
      if (a)
        return a.body.classList.add(t), function() {
          a.body.classList.remove(t);
        };
    }
  }, [t, e, r]), r;
}
export {
  L as FabricBase
};
