// @fluentui/react/lib/components/ComboBox/ComboBox.classNames.js
import { memoizeFunction as C } from "../../../../utilities/lib/memoize.js";
import { mergeStyles as r } from "../../../../merge-styles/lib/mergeStyles.js";
import "../../../../style-utilities/lib/index.js";
var c = /* @__PURE__ */ C(function(o, m, a, n, p, _, e, i) {
  return {
    container: r(o.__shadowConfig__, "ms-ComboBox-container", m, o.container),
    label: r(o.__shadowConfig__, o.label, n && o.labelDisabled),
    root: r(o.__shadowConfig__, "ms-ComboBox", i ? o.rootError : a && "is-open", p && "is-required", o.root, !e && o.rootDisallowFreeForm, i && !_ ? o.rootError : !n && _ && o.rootFocused, !n && {
      selectors: {
        ":hover": i ? o.rootError : !a && !_ && o.rootHovered,
        ":active": i ? o.rootError : o.rootPressed,
        ":focus": i ? o.rootError : o.rootFocused
      }
    }, n && ["is-disabled", o.rootDisabled]),
    input: r(o.__shadowConfig__, "ms-ComboBox-Input", o.input, n && o.inputDisabled),
    errorMessage: r(o.__shadowConfig__, o.errorMessage),
    callout: r(o.__shadowConfig__, "ms-ComboBox-callout", o.callout),
    optionsContainerWrapper: r(o.__shadowConfig__, "ms-ComboBox-optionsContainerWrapper", o.optionsContainerWrapper),
    optionsContainer: r(o.__shadowConfig__, "ms-ComboBox-optionsContainer", o.optionsContainer),
    header: r(o.__shadowConfig__, "ms-ComboBox-header", o.header),
    divider: r(o.__shadowConfig__, "ms-ComboBox-divider", o.divider),
    screenReaderText: r(o.__shadowConfig__, o.screenReaderText)
  };
}), f = /* @__PURE__ */ C(function(o) {
  return {
    optionText: r(o.__shadowConfig__, "ms-ComboBox-optionText", o.optionText),
    root: r(o.__shadowConfig__, "ms-ComboBox-option", o.root, {
      selectors: {
        ":hover": o.rootHovered,
        ":focus": o.rootFocused,
        ":active": o.rootPressed
      }
    }),
    optionTextWrapper: r(o.__shadowConfig__, o.optionTextWrapper)
  };
});
export {
  c as getClassNames,
  f as getComboBoxOptionClassNames
};
