// axios/lib/helpers/isURLSameOrigin.js
import c from "../utils.js";
import o from "../platform/index.js";
const f = o.hasStandardBrowserEnv ? (
  // Standard browser envs have full support of the APIs needed to test
  // whether the request URL is of the same origin as current location.
  /* @__PURE__ */ function() {
    const n = o.navigator && /(msie|trident)/i.test(o.navigator.userAgent), t = document.createElement("a");
    let e;
    function a(s) {
      let r = s;
      return n && (t.setAttribute("href", r), r = t.href), t.setAttribute("href", r), {
        href: t.href,
        protocol: t.protocol ? t.protocol.replace(/:$/, "") : "",
        host: t.host,
        search: t.search ? t.search.replace(/^\?/, "") : "",
        hash: t.hash ? t.hash.replace(/^#/, "") : "",
        hostname: t.hostname,
        port: t.port,
        pathname: t.pathname.charAt(0) === "/" ? t.pathname : "/" + t.pathname
      };
    }
    return e = a(window.location.href), function(r) {
      const i = c.isString(r) ? a(r) : r;
      return i.protocol === e.protocol && i.host === e.host;
    };
  }()
) : (
  // Non standard browser envs (web workers, react-native) lack needed support.
  /* @__PURE__ */ function() {
    return function() {
      return !0;
    };
  }()
);
export {
  f as default
};
