// @fluentui/react/lib/components/Callout/CalloutContent.base.js
import { __assign as k } from "../../../../../tslib/tslib.es6.js";
import * as s from "react";
import { DirectionalHint as ce } from "../../common/DirectionalHint.js";
import { getBoundsFromTargetWindow as ve, getRectangleFromTarget as fe, calculateGapSpace as me, positionCard as he, positionCallout as pe } from "../../utilities/positioning/positioning.js";
import { useWindowEx as ge } from "../../utilities/dom.js";
import { RectangleEdge as I } from "../../utilities/positioning/positioning.types.js";
import { AnimationClassNames as w } from "../../../../style-utilities/lib/classNames/AnimationClassNames.js";
import "../../../../style-utilities/lib/index.js";
import { getPropsWithDefaults as be } from "../../../../utilities/lib/getPropsWithDefaults.js";
import { useTarget as ye } from "../../../../react-hooks/lib/useTarget.js";
import { getNativeProps as Oe, divProperties as Pe } from "../../../../utilities/lib/properties.js";
import { css as He } from "../../../../utilities/lib/css.js";
import { Popup as Fe } from "../Popup/Popup.js";
import { shallowCompare as ke } from "../../../../utilities/lib/object.js";
import { classNamesFunction as De } from "../../../../utilities/lib/classNamesFunction.js";
import { focusFirstChild as Se } from "../../../../utilities/lib/focus.js";
import { useConst as Ne } from "../../../../react-hooks/lib/useConst.js";
import { on as E } from "../../../../utilities/lib/dom/on.js";
import { elementContains as Q } from "../../../../dom-utilities/lib/elementContains.js";
import { useMergedRefs as X } from "../../../../react-hooks/lib/useMergedRefs.js";
import { useAsync as U } from "../../../../react-hooks/lib/useAsync.js";
import { useOnEvent as Ce } from "../../../../react-hooks/lib/useOnEvent.js";
var L, Re = "CalloutContentBase", Me = (L = {}, L[I.top] = w.slideUpIn10, L[I.bottom] = w.slideDownIn10, L[I.left] = w.slideLeftIn10, L[I.right] = w.slideRightIn10, L), Z = { top: 0, left: 0 }, xe = {
  opacity: 0,
  filter: "opacity(0)",
  pointerEvents: "none"
}, Te = ["role", "aria-roledescription"], ee = {
  preventDismissOnLostFocus: !1,
  preventDismissOnScroll: !1,
  preventDismissOnResize: !1,
  isBeakVisible: !0,
  beakWidth: 16,
  gapSpace: 0,
  minPagePadding: 8,
  directionalHint: ce.bottomAutoEdge
}, Le = /* @__PURE__ */ De({
  disableCaching: !0
  // disabling caching because stylesProp.position mutates often
});
function Ae(i, r, e) {
  var t = i.bounds, o = i.minPagePadding, a = o === void 0 ? ee.minPagePadding : o, u = i.target, d = s.useState(!1), b = d[0], h = d[1], p = s.useRef(), y = s.useCallback(function() {
    if (!p.current || b) {
      var l = typeof t == "function" ? e ? t(u, e) : void 0 : t;
      !l && e && (l = ve(r.current, e), l = {
        top: l.top + a,
        left: l.left + a,
        right: l.right - a,
        bottom: l.bottom - a,
        width: l.width - a * 2,
        height: l.height - a * 2
      }), p.current = l, b && h(!1);
    }
    return p.current;
  }, [t, a, u, r, e, b]), c = U();
  return Ce(e, "resize", c.debounce(function() {
    h(!0);
  }, 500, { leading: !0 })), y;
}
function Ee(i, r, e, t) {
  var o, a = i.calloutMaxHeight, u = i.finalHeight, d = i.directionalHint, b = i.directionalHintFixed, h = i.hidden, p = i.gapSpace, y = i.beakWidth, c = i.isBeakVisible, l = i.coverTarget, D = s.useState(), R = D[0], O = D[1], M = (o = t == null ? void 0 : t.elementPosition) !== null && o !== void 0 ? o : {}, S = M.top, H = M.bottom, v = e != null && e.current ? fe(e.current) : void 0;
  return s.useEffect(function() {
    var N, P = (N = r()) !== null && N !== void 0 ? N : {}, n = P.top, g = P.bottom, f;
    (t == null ? void 0 : t.targetEdge) === I.top && (v != null && v.top) && !l && (g = v.top - me(c, y, p)), typeof S == "number" && g ? f = g - S : typeof H == "number" && typeof n == "number" && g && (f = g - n - H), !a && !h || a && f && a > f ? O(f) : O(a || void 0);
  }, [
    H,
    a,
    u,
    d,
    b,
    r,
    h,
    t,
    S,
    p,
    y,
    c,
    v,
    l
  ]), R;
}
function Ie(i, r, e, t, o, a) {
  var u = s.useState(), d = u[0], b = u[1], h = s.useRef(0), p = s.useRef(), y = U(), c = i.hidden, l = i.target, D = i.finalHeight, R = i.calloutMaxHeight, O = i.onPositioned, M = i.directionalHint, S = i.hideOverflow, H = i.preferScrollResizePositioning, v = ge(), N = s.useRef(), P;
  N.current !== a.current && (N.current = a.current, P = a.current ? v == null ? void 0 : v.getComputedStyle(a.current) : void 0);
  var n = P == null ? void 0 : P.overflowY;
  return s.useEffect(function() {
    if (c)
      b(void 0), h.current = 0;
    else {
      var g = y.requestAnimationFrame(function() {
        var f, F;
        if (r.current && e) {
          var T = k(k({}, i), { target: t.current, bounds: o() }), C = e.cloneNode(!0);
          C.style.maxHeight = R ? "".concat(R) : "", C.style.visibility = "hidden", (f = e.parentElement) === null || f === void 0 || f.appendChild(C);
          var B = p.current === l ? d : void 0, z = S || n === "clip" || n === "hidden", W = H && !z, x = D ? he(T, r.current, C, B, v) : pe(T, r.current, C, B, W, void 0, v);
          (F = e.parentElement) === null || F === void 0 || F.removeChild(C), !d && x || d && x && !we(d, x) && h.current < 5 ? (h.current++, b(x)) : h.current > 0 && (h.current = 0, O == null || O(d));
        }
      }, e);
      return p.current = l, function() {
        y.cancelAnimationFrame(g), p.current = void 0;
      };
    }
  }, [
    c,
    M,
    y,
    e,
    R,
    r,
    t,
    D,
    o,
    O,
    d,
    i,
    l,
    S,
    H,
    n,
    v
  ]), d;
}
function Be(i, r, e) {
  var t = i.hidden, o = i.setInitialFocus, a = U(), u = !!r;
  s.useEffect(function() {
    if (!t && o && u && e) {
      var d = a.requestAnimationFrame(function() {
        return Se(e);
      }, e);
      return function() {
        return a.cancelAnimationFrame(d);
      };
    }
  }, [t, u, a, e, o]);
}
function ze(i, r, e, t, o) {
  var a = i.hidden, u = i.onDismiss, d = i.preventDismissOnScroll, b = i.preventDismissOnResize, h = i.preventDismissOnLostFocus, p = i.dismissOnTargetClick, y = i.shouldDismissOnWindowFocus, c = i.preventDismissOnEvent, l = s.useRef(!1), D = U(), R = Ne([
    function() {
      l.current = !0;
    },
    function() {
      l.current = !1;
    }
  ]), O = !!r;
  return s.useEffect(function() {
    var M = function(n) {
      O && !d && v(n);
    }, S = function(n) {
      !b && !(c && c(n)) && (u == null || u(n));
    }, H = function(n) {
      h || v(n);
    }, v = function(n) {
      var g = n.composedPath ? n.composedPath() : [], f = g.length > 0 ? g[0] : n.target, F = e.current && !Q(e.current, f);
      if (F && l.current) {
        l.current = !1;
        return;
      }
      if (!t.current && F || n.target !== o && F && (!t.current || "stopPropagation" in t.current || p || f !== t.current && !Q(t.current, f))) {
        if (c && c(n))
          return;
        u == null || u(n);
      }
    }, N = function(n) {
      y && (c && !c(n) || !c && !h) && !(o != null && o.document.hasFocus()) && n.relatedTarget === null && (u == null || u(n));
    }, P = new Promise(function(n) {
      D.setTimeout(function() {
        if (!a && o) {
          var g = [
            E(o, "scroll", M, !0),
            E(o, "resize", S, !0),
            E(o.document.documentElement, "focus", H, !0),
            E(o.document.documentElement, "click", H, !0),
            E(o, "blur", N, !0)
          ];
          n(function() {
            g.forEach(function(f) {
              return f();
            });
          });
        }
      }, 0);
    });
    return function() {
      P.then(function(n) {
        return n();
      });
    };
  }, [
    a,
    D,
    e,
    t,
    o,
    u,
    y,
    p,
    h,
    b,
    d,
    O,
    c
  ]), R;
}
var We = /* @__PURE__ */ s.memo(s.forwardRef(function(i, r) {
  var e = be(ee, i), t = e.styles, o = e.style, a = e.ariaLabel, u = e.ariaDescribedBy, d = e.ariaLabelledBy, b = e.className, h = e.isBeakVisible, p = e.children, y = e.beakWidth, c = e.calloutWidth, l = e.calloutMaxWidth, D = e.calloutMinWidth, R = e.doNotLayer, O = e.finalHeight, M = e.hideOverflow, S = M === void 0 ? !!O : M, H = e.backgroundColor, v = e.calloutMaxHeight, N = e.onScroll, P = e.shouldRestoreFocus, n = P === void 0 ? !0 : P, g = e.target, f = e.hidden, F = e.onLayerMounted, T = e.popupProps, C = s.useRef(null), B = s.useRef(null), z = X(B, T == null ? void 0 : T.ref), W = s.useState(null), x = W[0], ie = W[1], te = s.useCallback(function(de) {
    ie(de);
  }, []), re = X(C, r), Y = ye(e.target, {
    current: x
  }), _ = Y[0], V = Y[1], q = Ae(e, _, V), m = Ie(e, C, x, _, q, z), ne = Ee(e, q, _, m), G = ze(e, m, C, _, V), oe = G[0], ae = G[1], se = (m == null ? void 0 : m.elementPosition.top) && (m == null ? void 0 : m.elementPosition.bottom), K = k(k({}, m == null ? void 0 : m.elementPosition), { maxHeight: ne });
  if (se && (K.bottom = void 0), Be(e, m, x), s.useEffect(function() {
    f || F == null || F();
  }, [f]), !V)
    return null;
  var j = S, J = h && !!g, A = Le(t, {
    theme: e.theme,
    className: b,
    overflowYHidden: j,
    calloutWidth: c,
    positions: m,
    beakWidth: y,
    backgroundColor: H,
    calloutMaxWidth: l,
    calloutMinWidth: D,
    doNotLayer: R
  }), le = k(k({ maxHeight: v || "100%" }, o), j && { overflowY: "hidden" }), ue = e.hidden ? { visibility: "hidden" } : void 0;
  return s.createElement(
    "div",
    { ref: re, className: A.container, style: ue },
    s.createElement(
      "div",
      k({}, Oe(e, Pe, Te), {
        className: He(A.root, m && m.targetEdge && Me[m.targetEdge]),
        style: m ? k({}, K) : xe,
        // Safari and Firefox on Mac OS requires this to back-stop click events so focus remains in the Callout.
        // See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
        tabIndex: -1,
        ref: te
      }),
      J && s.createElement("div", { className: A.beak, style: _e(m) }),
      J && s.createElement("div", { className: A.beakCurtain }),
      s.createElement(
        Fe,
        k({
          // don't use getNativeElementProps for role and roledescription because it will also
          // pass through data-* props (resulting in them being used in two places)
          role: e.role,
          "aria-roledescription": e["aria-roledescription"],
          ariaDescribedBy: u,
          ariaLabel: a,
          ariaLabelledBy: d,
          className: A.calloutMain,
          onDismiss: e.onDismiss,
          onMouseDown: oe,
          onMouseUp: ae,
          onRestoreFocus: e.onRestoreFocus,
          onScroll: N,
          shouldRestoreFocus: n,
          style: le
        }, T, { ref: z }),
        p
      )
    )
  );
}), function(i, r) {
  return !r.shouldUpdateWhenHidden && i.hidden && r.hidden ? !0 : ke(i, r);
});
function _e(i) {
  var r, e, t = k(k({}, (r = i == null ? void 0 : i.beakPosition) === null || r === void 0 ? void 0 : r.elementPosition), { display: !((e = i == null ? void 0 : i.beakPosition) === null || e === void 0) && e.hideBeak ? "none" : void 0 });
  return !t.top && !t.bottom && !t.left && !t.right && (t.left = Z.left, t.top = Z.top), t;
}
function we(i, r) {
  return $(i.elementPosition, r.elementPosition) && $(i.beakPosition.elementPosition, r.beakPosition.elementPosition);
}
function $(i, r) {
  for (var e in r)
    if (r.hasOwnProperty(e)) {
      var t = i[e], o = r[e];
      if (t !== void 0 && o !== void 0) {
        if (t.toFixed(2) !== o.toFixed(2))
          return !1;
      } else
        return !1;
    }
  return !0;
}
We.displayName = Re;
export {
  We as CalloutContentBase
};
