// @fluentui/react/lib/components/ContextualMenu/ContextualMenu.base.js
import { __spreadArray as Ar, __rest as We, __assign as S } from "../../../../../tslib/tslib.es6.js";
import * as u from "react";
import { ContextualMenuItemType as N } from "./ContextualMenu.types.js";
import { DirectionalHint as ce } from "../../common/DirectionalHint.js";
import { ContextualMenuItem as Or } from "./ContextualMenuItem.js";
import { getItemStyles as Hr } from "./ContextualMenu.classNames.js";
import { memoizeFunction as Kr } from "../../../../utilities/lib/memoize.js";
import { concatStyleSetsWithProps as Lr } from "../../../../merge-styles/lib/concatStyleSetsWithProps.js";
import "../../../../style-utilities/lib/index.js";
import { getPropsWithDefaults as wr } from "../../../../utilities/lib/getPropsWithDefaults.js";
import { useResponsiveMode as xr } from "../../utilities/hooks/useResponsiveMode.js";
import { FocusZoneDirection as le, FocusZoneTabbableElements as Br } from "../../../../react-focus/lib/components/FocusZone/FocusZone.types.js";
import { css as Ae } from "../../../../utilities/lib/css.js";
import { ResponsiveMode as Wr } from "../../utilities/decorators/withResponsiveMode.js";
import { MenuContext as _r } from "../../utilities/MenuContext/MenuContext.js";
import { FocusRects as Ur } from "../../../../utilities/lib/useFocusRects.js";
import { shallowCompare as Zr, assign as Vr } from "../../../../utilities/lib/object.js";
import { classNamesFunction as _e } from "../../../../utilities/lib/classNamesFunction.js";
import { useIsomorphicLayoutEffect as zr } from "../../../../utilities/lib/useIsomorphicLayoutEffect.js";
import { getDocument as jr } from "../../../../utilities/lib/dom/getDocument.js";
import { composeComponentAs as G } from "../../../../utilities/lib/componentAs/composeComponentAs.js";
import { getNativeProps as Yr, divProperties as qr } from "../../../../utilities/lib/properties.js";
import { getRTL as de } from "../../../../utilities/lib/rtl.js";
import { KeyCodes as L } from "../../../../utilities/lib/KeyCodes.js";
import { isMac as Oe } from "../../../../utilities/lib/osDetector.js";
import { isIOS as He } from "../../../../utilities/lib/mobileDetector.js";
import { getLastFocusable as Gr, getFirstFocusable as Jr, shouldWrapFocus as Qr } from "../../../../utilities/lib/focus.js";
import { FocusZone as Xr } from "../../../../react-focus/lib/components/FocusZone/FocusZone.js";
import { ContextualMenuAnchor as $r } from "./ContextualMenuItemWrapper/ContextualMenuAnchor.js";
import { ContextualMenuSplitButton as en } from "./ContextualMenuItemWrapper/ContextualMenuSplitButton.js";
import { ContextualMenuButton as rn } from "./ContextualMenuItemWrapper/ContextualMenuButton.js";
import { useAsync as nn } from "../../../../react-hooks/lib/useAsync.js";
import { useId as Ue } from "../../../../react-hooks/lib/useId.js";
import { useWarnings as tn } from "../../../../react-hooks/lib/useWarnings.js";
import { useTarget as on } from "../../../../react-hooks/lib/useTarget.js";
import { Callout as un } from "../Callout/Callout.js";
import { usePrevious as an } from "../../../../react-hooks/lib/usePrevious.js";
import { isItemDisabled as Ke, getIsChecked as Le, hasSubmenu as fe } from "../../utilities/contextualMenu/contextualMenuUtility.js";
var sn = /* @__PURE__ */ _e(), cn = /* @__PURE__ */ _e(), ln = {
  items: [],
  shouldFocusOnMount: !0,
  gapSpace: 0,
  directionalHint: ce.bottomAutoEdge,
  beakWidth: 16
};
function we(n) {
  for (var t = 0, o = 0, c = n; o < c.length; o++) {
    var e = c[o];
    if (e.itemType !== N.Divider && e.itemType !== N.Header) {
      var s = e.customOnRenderListLength ? e.customOnRenderListLength : 1;
      t += s;
    }
  }
  return t;
}
function Ze(n, t) {
  var o = t == null ? void 0 : t.target, c = n.subMenuProps ? n.subMenuProps.items : n.items;
  if (c) {
    for (var e = [], s = 0, y = c; s < y.length; s++) {
      var f = y[s];
      if (f.preferMenuTargetAsEventTarget) {
        var v = f.onClick, p = We(f, ["onClick"]);
        e.push(S(S({}, p), { onClick: je(v, o) }));
      } else
        e.push(f);
    }
    return e;
  }
}
function dn(n) {
  return n.some(function(t) {
    return !!(t.canCheck || t.sectionProps && t.sectionProps.items.some(function(o) {
      return o.canCheck === !0;
    }));
  });
}
var Ve = 250, me = "ContextualMenu", fn = /* @__PURE__ */ Kr(function() {
  for (var n = [], t = 0; t < arguments.length; t++)
    n[t] = arguments[t];
  return function(o) {
    return Lr.apply(void 0, Ar([o, Hr], n, !1));
  };
});
function mn(n, t) {
  var o = n.hidden, c = o === void 0 ? !1 : o, e = n.onMenuDismissed, s = n.onMenuOpened, y = an(c), f = u.useRef(s), v = u.useRef(e), p = u.useRef(n);
  f.current = s, v.current = e, p.current = n, u.useEffect(function() {
    var k, E;
    c && y === !1 ? (k = v.current) === null || k === void 0 || k.call(v, p.current) : !c && y !== !1 && ((E = f.current) === null || E === void 0 || E.call(f, p.current));
  }, [c, y]), u.useEffect(function() {
    return function() {
      var k;
      return (k = v.current) === null || k === void 0 ? void 0 : k.call(v, p.current);
    };
  }, []);
}
function vn(n, t) {
  var o = n.hidden, c = n.items, e = n.theme, s = n.className, y = n.id, f = n.target, v = u.useState(), p = v[0], k = v[1], E = u.useState(), _ = E[0], A = E[1], O = u.useState(), B = O[0], W = O[1], U = Ue(me, y), H = u.useCallback(function() {
    W(void 0), k(void 0), A(void 0);
  }, []), j = u.useCallback(function(I, a, d) {
    var g = I.key;
    p !== g && (a.focus(), W(d), k(g), A(a));
  }, [p]);
  u.useEffect(function() {
    o && H();
  }, [o, H]);
  var i = yn(t, H), b = function() {
    var I = ze(p, c), a = null;
    if (I && (a = {
      items: Ze(I, { target: f }),
      target: _,
      onDismiss: i,
      isSubMenu: !0,
      id: U,
      shouldFocusOnMount: !0,
      shouldFocusOnContainer: B,
      directionalHint: de(e) ? ce.leftTopEdge : ce.rightTopEdge,
      className: s,
      gapSpace: 0,
      isBeakVisible: !1
    }, I.subMenuProps && Vr(a, I.subMenuProps), I.preferMenuTargetAsEventTarget)) {
      var d = I.onItemClick;
      a.onItemClick = je(d, f);
    }
    return a;
  };
  return [p, j, b, i];
}
function pn(n) {
  var t = n.delayUpdateFocusOnHover, o = n.hidden, c = u.useRef(!t), e = u.useRef(!1);
  u.useEffect(function() {
    c.current = !t, e.current = o ? !1 : !t && e.current;
  }, [t, o]);
  var s = u.useCallback(function() {
    t && (c.current = !1);
  }, [t]);
  return [c, e, s];
}
function hn(n, t, o) {
  var c = n.hidden, e = n.onRestoreFocus, s = u.useRef(), y = u.useCallback(function(f) {
    var v, p;
    e ? e(f) : f != null && f.documentContainsFocus && ((p = (v = s.current) === null || v === void 0 ? void 0 : v.focus) === null || p === void 0 || p.call(v));
  }, [e]);
  return zr(function() {
    var f, v;
    if (c)
      s.current && (y({
        originalElement: s.current,
        containsFocus: !0,
        documentContainsFocus: ((v = jr()) === null || v === void 0 ? void 0 : v.hasFocus()) || !1
      }), s.current = void 0);
    else {
      var p = t == null ? void 0 : t.document.activeElement;
      !(!((f = o.current) === null || f === void 0) && f.contains(p)) && p.tagName !== "BODY" && (s.current = p);
    }
  }, [c, t == null ? void 0 : t.document.activeElement, y, o]), [y];
}
function Mn(n, t, o, c) {
  var e = n.theme, s = n.isSubMenu, y = n.focusZoneProps, f = y === void 0 ? {} : y, v = f.checkForNoWrap, p = f.direction, k = p === void 0 ? le.vertical : p, E = u.useRef(), _ = function(i, b, I) {
    var a = !1;
    return b(i) && (t(i, I), i.preventDefault(), i.stopPropagation(), a = !0), a;
  }, A = function(i) {
    var b = de(e) ? L.right : L.left;
    return i.which !== b || !s ? !1 : !!(k === le.vertical || v && !Qr(i.target, "data-no-horizontal-wrap"));
  }, O = function(i) {
    return (
      // eslint-disable-next-line deprecation/deprecation
      i.which === L.escape || A(i) || // eslint-disable-next-line deprecation/deprecation
      i.which === L.up && (i.altKey || i.metaKey)
    );
  }, B = function(i) {
    E.current = xe(i);
    var b = i.which === L.escape && (Oe() || He());
    return _(i, O, b);
  }, W = function(i) {
    var b = E.current && xe(i);
    return E.current = !1, !!b && !(He() || Oe());
  }, U = function(i) {
    return _(
      i,
      W,
      !0
      /* dismissAllMenus */
    );
  }, H = function(i) {
    var b = B(i);
    if (!(b || !o.current)) {
      var I = !!(i.altKey || i.metaKey), a = i.which === L.up, d = i.which === L.down;
      if (!I && (a || d)) {
        var g = a ? Gr(o.current, o.current.lastChild, !0) : Jr(o.current, o.current.firstChild, !0);
        g && (g.focus(), i.preventDefault(), i.stopPropagation());
      }
    }
  }, j = function(i, b) {
    var I = de(e) ? L.left : L.right;
    !i.disabled && // eslint-disable-next-line deprecation/deprecation
    (b.which === I || b.which === L.enter || b.which === L.down && (b.altKey || b.metaKey)) && (c(i, b.currentTarget), b.preventDefault());
  };
  return [B, U, H, j];
}
function gn(n) {
  var t = u.useRef(!0), o = u.useRef(), c = function() {
    !t.current && o.current !== void 0 ? (n.clearTimeout(o.current), o.current = void 0) : t.current = !1, o.current = n.setTimeout(function() {
      t.current = !0;
    }, Ve);
  };
  return [c, t];
}
function yn(n, t) {
  var o = u.useRef(!1);
  u.useEffect(function() {
    return o.current = !0, function() {
      o.current = !1;
    };
  }, []);
  var c = function(e, s) {
    s ? n(e, s) : o.current && t();
  };
  return c;
}
function bn(n, t) {
  var o = n.subMenuHoverDelay, c = o === void 0 ? Ve : o, e = u.useRef(void 0), s = function() {
    e.current !== void 0 && (t.clearTimeout(e.current), e.current = void 0);
  }, y = function(f) {
    e.current = t.setTimeout(function() {
      f(), s();
    }, c);
  };
  return [s, y, e];
}
function In(n, t, o, c, e, s, y, f, v, p, k, E, _) {
  var A = n.target, O = function(a, d, g) {
    e.current && (s.current = !0), !W() && H(a, d, g);
  }, B = function(a, d, g) {
    var P = d.currentTarget;
    if (e.current)
      s.current = !0;
    else
      return;
    !t.current || o.current !== void 0 || P === (c == null ? void 0 : c.document.activeElement) || H(a, d, g);
  }, W = function() {
    return !t.current || !s.current;
  }, U = function(a, d) {
    var g;
    if (!W() && (p(), y === void 0))
      if (f.current.setActive)
        try {
          f.current.setActive();
        } catch {
        }
      else
        (g = f.current) === null || g === void 0 || g.focus();
  }, H = function(a, d, g) {
    var P = g || d.currentTarget;
    a.key !== y && (p(), y === void 0 && P.focus(), fe(a) ? (d.stopPropagation(), v(function() {
      P.focus(), k(a, P, !0);
    })) : v(function() {
      E(d), P.focus();
    }));
  }, j = function(a, d) {
    i(a, d, d.currentTarget);
  }, i = function(a, d, g) {
    var P = Ze(a, { target: A });
    if (p(), !fe(a) && (!P || !P.length))
      I(a, d);
    else if (a.key !== y) {
      var ee = typeof n.shouldFocusOnContainer == "boolean" ? n.shouldFocusOnContainer : d.nativeEvent.pointerType === "mouse";
      k(a, g, ee);
    }
    d.stopPropagation(), d.preventDefault();
  }, b = function(a, d) {
    I(a, d), d.stopPropagation();
  }, I = function(a, d) {
    if (!(a.disabled || a.isDisabled)) {
      a.preferMenuTargetAsEventTarget && Ye(d, A);
      var g = !1;
      a.onClick ? g = !!a.onClick(d, a) : n.onItemClick && (g = !!n.onItemClick(d, a)), (g || !d.defaultPrevented) && _(d, !0);
    }
  };
  return [
    O,
    B,
    U,
    j,
    b,
    I,
    i
  ];
}
var Cn = /* @__PURE__ */ u.memo(u.forwardRef(function(n, t) {
  var o, c = wr(ln, n);
  c.ref;
  var e = We(c, ["ref"]), s = u.useRef(null), y = nn(), f = Ue(me, e.id);
  tn({
    name: me,
    props: e,
    deprecations: {
      getMenuClassNames: "styles"
    }
  });
  var v = function(r, l) {
    var h;
    return (h = e.onDismiss) === null || h === void 0 ? void 0 : h.call(e, r, l);
  }, p = on(e.target, s), k = p[0], E = p[1], _ = hn(e, E, s)[0], A = vn(e, v), O = A[0], B = A[1], W = A[2], U = A[3], H = pn(e), j = H[0], i = H[1], b = H[2], I = gn(y), a = I[0], d = I[1], g = bn(e, y), P = g[0], ee = g[1], qe = g[2], Ge = xr(s, e.responsiveMode);
  mn(e);
  var J = Mn(e, v, s, B), Je = J[0], ve = J[1], Qe = J[2], Xe = J[3], Y = In(e, d, qe, E, j, i, O, s, ee, P, B, U, v), $e = Y[0], er = Y[1], rr = Y[2], re = Y[3], nr = Y[4], tr = Y[5], pe = Y[6], he = function(r, l, h) {
    var M = 0, R = r.items, F = r.totalItemCount, m = r.hasCheckmarks, D = r.hasIcons;
    return u.createElement("ul", { className: l.list, onKeyDown: Je, onKeyUp: ve, role: "presentation" }, R.map(function(C, T) {
      var K = ne(C, T, M, F, m, D, l);
      if (C.itemType !== N.Divider && C.itemType !== N.Header) {
        var w = C.customOnRenderListLength ? C.customOnRenderListLength : 1;
        M += w;
      }
      return K;
    }));
  }, or = function(r, l) {
    var h = e.focusZoneAs, M = h === void 0 ? Xr : h;
    return u.createElement(M, S({}, l), r);
  }, ne = function(r, l, h, M, R, F, m) {
    var D, C = [], T = r.iconProps || { iconName: "None" }, K = r.getItemClassNames, w = r.itemProps, $ = w ? w.styles : void 0, q = r.itemType === N.Divider ? r.className : void 0, z = r.submenuIconProps ? r.submenuIconProps.className : "", x;
    if (K)
      x = K(e.theme, Ke(r), O === r.key, !!Le(r), !!r.href, T.iconName !== "None", r.className, q, T.className, z, r.primaryDisabled);
    else {
      var se = {
        theme: e.theme,
        disabled: Ke(r),
        expanded: O === r.key,
        checked: !!Le(r),
        isAnchorLink: !!r.href,
        knownIcon: T.iconName !== "None",
        itemClassName: r.className,
        dividerClassName: q,
        iconClassName: T.className,
        subMenuClassName: z,
        primaryDisabled: r.primaryDisabled
      };
      x = cn(fn((D = m.subComponentStyles) === null || D === void 0 ? void 0 : D.menuItem, $), se);
    }
    switch ((r.text === "-" || r.name === "-") && (r.itemType = N.Divider), r.itemType) {
      case N.Divider:
        C.push(Q(l, x));
        break;
      case N.Header:
        C.push(Q(l, x));
        var ie = Me(r, x, m, l, R, F);
        C.push(te(ie, r.key || l, x, r.title));
        break;
      case N.Section:
        C.push(ar(r, x, m, l, R, F));
        break;
      default:
        var Pe = function() {
          return sr(r, x, l, h, M, R, F);
        }, Pr = e.onRenderContextualMenuItem ? e.onRenderContextualMenuItem(r, Pe) : Pe();
        C.push(te(Pr, r.key || l, x, r.title));
        break;
    }
    return u.createElement(u.Fragment, { key: r.key }, C);
  }, ur = function(r, l) {
    var h = r.index, M = r.focusableElementIndex, R = r.totalItemCount, F = r.hasCheckmarks, m = r.hasIcons;
    return ne(r, h, M, R, F, m, l);
  }, ar = function(r, l, h, M, R, F) {
    var m = r.sectionProps;
    if (m) {
      var D, C;
      if (m.title) {
        var T = void 0, K = "";
        if (typeof m.title == "string") {
          var w = f + m.title.replace(/\s/g, "");
          T = {
            key: "section-".concat(m.title, "-title"),
            itemType: N.Header,
            text: m.title,
            id: w
          }, K = w;
        } else {
          var $ = m.title.id || f + m.title.key.replace(/\s/g, "");
          T = S(S({}, m.title), { id: $ }), K = $;
        }
        T && (C = {
          role: "group",
          "aria-labelledby": K
        }, D = Me(T, l, h, M, R, F));
      }
      if (m.items && m.items.length > 0) {
        var q = 0;
        return u.createElement(
          "li",
          { role: "presentation", key: m.key || r.key || "section-".concat(M) },
          u.createElement(
            "div",
            S({}, C),
            u.createElement(
              "ul",
              { className: h.list, role: "presentation" },
              m.topDivider && Q(M, l, !0, !0),
              D && te(D, r.key || M, l, r.title),
              m.items.map(function(z, x) {
                var se = ne(z, x, q, we(m.items), R, F, h);
                if (z.itemType !== N.Divider && z.itemType !== N.Header) {
                  var ie = z.customOnRenderListLength ? z.customOnRenderListLength : 1;
                  q += ie;
                }
                return se;
              }),
              m.bottomDivider && Q(M, l, !1, !0)
            )
          )
        );
      }
    }
  }, te = function(r, l, h, M) {
    return u.createElement("li", { role: "presentation", title: M, key: l, className: h.item }, r);
  }, Q = function(r, l, h, M) {
    return M || r > 0 ? u.createElement("li", { role: "separator", key: "separator-" + r + (h === void 0 ? "" : h ? "-top" : "-bottom"), className: l.divider, "aria-hidden": "true" }) : null;
  }, sr = function(r, l, h, M, R, F, m) {
    if (r.onRender)
      return r.onRender(S({ "aria-posinset": M + 1, "aria-setsize": R }, r), v);
    var D = e.contextualMenuItemAs, C = {
      item: r,
      classNames: l,
      index: h,
      focusableElementIndex: M,
      totalItemCount: R,
      hasCheckmarks: F,
      hasIcons: m,
      contextualMenuItemAs: D,
      onItemMouseEnter: $e,
      onItemMouseLeave: rr,
      onItemMouseMove: er,
      onItemMouseDown: kn,
      executeItemClick: tr,
      onItemKeyDown: Xe,
      expandedMenuItemKey: O,
      openSubMenu: B,
      dismissSubMenu: U,
      dismissMenu: v
    };
    if (r.href) {
      var T = $r;
      return r.contextualMenuItemWrapperAs && (T = G(r.contextualMenuItemWrapperAs, T)), u.createElement(T, S({}, C, { onItemClick: nr }));
    }
    if (r.split && fe(r)) {
      var K = en;
      return r.contextualMenuItemWrapperAs && (K = G(r.contextualMenuItemWrapperAs, K)), u.createElement(K, S({}, C, { onItemClick: re, onItemClickBase: pe, onTap: P }));
    }
    var w = rn;
    return r.contextualMenuItemWrapperAs && (w = G(r.contextualMenuItemWrapperAs, w)), u.createElement(w, S({}, C, { onItemClick: re, onItemClickBase: pe }));
  }, Me = function(r, l, h, M, R, F) {
    var m = Or;
    r.contextualMenuItemAs && (m = G(r.contextualMenuItemAs, m)), e.contextualMenuItemAs && (m = G(e.contextualMenuItemAs, m));
    var D = r.itemProps, C = r.id, T = D && Yr(D, qr);
    return (
      // eslint-disable-next-line deprecation/deprecation
      u.createElement(
        "div",
        S({ id: C, className: h.header }, T, { style: r.style }),
        u.createElement(m, S({ item: r, classNames: l, index: M, onCheckmarkClick: R ? re : void 0, hasIcons: F }, D))
      )
    );
  }, X = e.isBeakVisible, Z = e.items, ge = e.labelElementId, ir = e.id, ye = e.className, cr = e.beakWidth, lr = e.directionalHint, dr = e.directionalHintForRTL, fr = e.alignTargetEdge, mr = e.gapSpace, vr = e.coverTarget, be = e.ariaLabel, pr = e.doNotLayer, hr = e.target, Mr = e.bounds, Ie = e.useTargetWidth, Ce = e.useTargetAsMinWidth, gr = e.directionalHintFixed, yr = e.shouldFocusOnMount, br = e.shouldFocusOnContainer, ke = e.title, Ir = e.styles, Te = e.theme, oe = e.calloutProps, Ee = e.onRenderSubMenu, Cr = Ee === void 0 ? Be : Ee, Re = e.onRenderMenuList, kr = Re === void 0 ? function(r, l) {
    return he(r, V);
  } : Re, Tr = e.focusZoneProps, De = e.getMenuClassNames, V = De ? De(Te, ye) : sn(Ir, {
    theme: Te,
    className: ye
  }), Er = Se(Z);
  function Se(r) {
    for (var l = 0, h = r; l < h.length; l++) {
      var M = h[l];
      if (M.iconProps || M.itemType === N.Section && M.sectionProps && Se(M.sectionProps.items))
        return !0;
    }
    return !1;
  }
  var Rr = S(S({ direction: le.vertical, handleTabKey: Br.all, isCircularNavigation: !0, "data-tabster": '{"uncontrolled": {}, "focusable": { "excludeFromMover": true }}' }, Tr), { className: Ae(V.root, (o = e.focusZoneProps) === null || o === void 0 ? void 0 : o.className) }), Dr = dn(Z), Fe = O && e.hidden !== !0 ? W() : null;
  X = X === void 0 ? Ge <= Wr.medium : X;
  var ue, ae = k.current;
  if ((Ie || Ce) && ae && ae.offsetWidth) {
    var Sr = ae.getBoundingClientRect(), Ne = Sr.width - 2;
    Ie ? ue = {
      width: Ne
    } : Ce && (ue = {
      minWidth: Ne
    });
  }
  if (Z && Z.length > 0) {
    var Fr = we(Z), Nr = V.subComponentStyles ? V.subComponentStyles.callout : void 0;
    return u.createElement(_r.Consumer, null, function(r) {
      return u.createElement(
        un,
        S({ styles: Nr, onRestoreFocus: _ }, oe, { target: hr || r.target, isBeakVisible: X, beakWidth: cr, directionalHint: lr, directionalHintForRTL: dr, gapSpace: mr, coverTarget: vr, doNotLayer: pr, className: Ae("ms-ContextualMenu-Callout", oe && oe.className), setInitialFocus: yr, onDismiss: e.onDismiss || r.onDismiss, onScroll: a, bounds: Mr, directionalHintFixed: gr, alignTargetEdge: fr, hidden: e.hidden || r.hidden, ref: t }),
        u.createElement(
          "div",
          { style: ue, ref: s, id: ir, className: V.container, tabIndex: br ? 0 : -1, onKeyDown: Qe, onKeyUp: ve, onFocusCapture: b, "aria-label": be, "aria-labelledby": ge, role: "menu" },
          ke && u.createElement(
            "div",
            { className: V.title },
            " ",
            ke,
            " "
          ),
          Z && Z.length ? or(kr({
            ariaLabel: be,
            items: Z,
            totalItemCount: Fr,
            hasCheckmarks: Dr,
            hasIcons: Er,
            defaultMenuItemRenderer: function(l) {
              return ur(l, V);
            },
            labelElementId: ge
          }, function(l, h) {
            return he(l, V);
          }), Rr) : null,
          Fe && Cr(Fe, Be)
        ),
        u.createElement(Ur, null)
      );
    });
  } else
    return null;
}), function(n, t) {
  return !t.shouldUpdateWhenHidden && n.hidden && t.hidden ? !0 : Zr(n, t);
});
Cn.displayName = "ContextualMenuBase";
function xe(n) {
  return n.which === L.alt || n.key === "Meta";
}
function kn(n, t) {
  var o;
  (o = n.onMouseDown) === null || o === void 0 || o.call(n, n, t);
}
function Be(n, t) {
  throw Error("ContextualMenuBase: onRenderSubMenu callback is null or undefined. Please ensure to set `onRenderSubMenu` property either manually or with `styled` helper.");
}
function ze(n, t) {
  for (var o = 0, c = t; o < c.length; o++) {
    var e = c[o];
    if (e.itemType === N.Section && e.sectionProps) {
      var s = ze(n, e.sectionProps.items);
      if (s)
        return s;
    } else if (e.key && e.key === n)
      return e;
  }
}
function je(n, t) {
  return n && function(o, c) {
    return Ye(o, t), n(o, c);
  };
}
function Ye(n, t) {
  n && t && (n.persist(), t instanceof Event ? n.target = t.target : t instanceof Element && (n.target = t));
}
export {
  Cn as ContextualMenuBase,
  dn as canAnyMenuItemsCheck,
  Ze as getSubmenuItems
};
