// @fluentui/react/lib/components/Button/ActionButton/ActionButton.styles.js
import { getStyles as i } from "../BaseButton.styles.js";
import { memoizeFunction as p } from "../../../../../utilities/lib/memoize.js";
import "../../../../../style-utilities/lib/index.js";
import { concatStyleSets as s } from "../../../../../merge-styles/lib/concatStyleSets.js";
import { HighContrastSelector as a } from "../../../../../style-utilities/lib/styles/CommonStyles.js";
var d = "40px", y = "0 4px", C = /* @__PURE__ */ p(function(r, l) {
  var o, t, e, n = i(r), c = {
    root: (o = {
      padding: y,
      height: d,
      color: r.palette.neutralPrimary,
      backgroundColor: "transparent",
      border: "1px solid transparent"
    }, o[a] = {
      borderColor: "Window"
    }, o),
    rootHovered: (t = {
      color: r.palette.themePrimary
    }, t[a] = {
      color: "Highlight"
    }, t),
    iconHovered: {
      color: r.palette.themePrimary
    },
    rootPressed: {
      color: r.palette.black
    },
    rootExpanded: {
      color: r.palette.themePrimary
    },
    iconPressed: {
      color: r.palette.themeDarker
    },
    rootDisabled: (e = {
      color: r.palette.neutralTertiary,
      backgroundColor: "transparent",
      borderColor: "transparent"
    }, e[a] = {
      color: "GrayText"
    }, e),
    rootChecked: {
      color: r.palette.black
    },
    iconChecked: {
      color: r.palette.themeDarker
    },
    flexContainer: {
      justifyContent: "flex-start"
    },
    icon: {
      color: r.palette.themeDarkAlt
    },
    iconDisabled: {
      color: "inherit"
    },
    menuIcon: {
      color: r.palette.neutralSecondary
    },
    textContainer: {
      flexGrow: 0
    }
  };
  return s(n, c, l);
});
export {
  C as getStyles
};
