// @fluentui/react/lib/components/Button/BaseButton.styles.js
import { memoizeFunction as f } from "../../../../utilities/lib/memoize.js";
import "../../../../style-utilities/lib/index.js";
import { hiddenContentStyle as x } from "../../../../style-utilities/lib/styles/hiddenContentStyle.js";
import { getFocusStyle as s } from "../../../../style-utilities/lib/styles/getFocusStyle.js";
import { HighContrastSelector as d } from "../../../../style-utilities/lib/styles/CommonStyles.js";
var c = {
  outline: 0
}, u = function(o) {
  return {
    fontSize: o,
    margin: "0 4px",
    height: "16px",
    lineHeight: "16px",
    textAlign: "center",
    flexShrink: 0
  };
}, v = /* @__PURE__ */ f(function(o) {
  var e, r, t = o.semanticColors, p = o.effects, i = o.fonts, b = t.buttonBorder, l = t.disabledBackground, n = t.disabledText, a = {
    left: -2,
    top: -2,
    bottom: -2,
    right: -2,
    outlineColor: "ButtonText"
  };
  return {
    root: [
      s(o, { inset: 1, highContrastStyle: a, borderColor: "transparent" }),
      o.fonts.medium,
      {
        border: "1px solid " + b,
        borderRadius: p.roundedCorner2,
        boxSizing: "border-box",
        cursor: "pointer",
        display: "inline-block",
        padding: "0 16px",
        textDecoration: "none",
        textAlign: "center",
        userSelect: "none",
        // IE11 workaround for preventing shift of child elements of a button when active.
        ":active > span": {
          position: "relative",
          left: 0,
          top: 0
        }
      }
    ],
    rootDisabled: [
      s(o, { inset: 1, highContrastStyle: a, borderColor: "transparent" }),
      {
        backgroundColor: l,
        borderColor: l,
        color: n,
        cursor: "default",
        ":hover": c,
        ":focus": c
      }
    ],
    iconDisabled: (e = {
      color: n
    }, e[d] = {
      color: "GrayText"
    }, e),
    menuIconDisabled: (r = {
      color: n
    }, r[d] = {
      color: "GrayText"
    }, r),
    flexContainer: {
      display: "flex",
      height: "100%",
      flexWrap: "nowrap",
      justifyContent: "center",
      alignItems: "center"
    },
    description: {
      display: "block"
    },
    textContainer: {
      flexGrow: 1,
      display: "block"
    },
    icon: u(i.mediumPlus.fontSize),
    menuIcon: u(i.small.fontSize),
    label: {
      margin: "0 4px",
      lineHeight: "100%",
      display: "block"
    },
    screenReaderText: x
  };
});
export {
  v as getStyles
};
