// @fluentui/react/lib/components/Label/Label.styles.js
import { __assign as m } from "../../../../../tslib/tslib.es6.js";
import "../../../../style-utilities/lib/index.js";
import { FontWeights as c } from "../../../../theme/lib/fonts/FluentFonts.js";
import { HighContrastSelector as g, getHighContrastNoAdjustStyle as h } from "../../../../style-utilities/lib/styles/CommonStyles.js";
var C = function(o) {
  var e, t = o.theme, a = o.className, l = o.disabled, i = o.required, r = t.semanticColors, d = c.semibold, s = r.bodyText, n = r.disabledBodyText, b = r.errorText;
  return {
    root: [
      "ms-Label",
      t.fonts.medium,
      {
        fontWeight: d,
        color: s,
        boxSizing: "border-box",
        boxShadow: "none",
        margin: 0,
        display: "block",
        padding: "5px 0",
        wordWrap: "break-word",
        overflowWrap: "break-word"
      },
      l && {
        color: n,
        selectors: (e = {}, e[g] = m({ color: "GrayText" }, h()), e)
      },
      i && {
        selectors: {
          "::after": {
            content: "' *'",
            color: b,
            paddingRight: 12
          }
        }
      },
      a
    ]
  };
};
export {
  C as getStyles
};
