// @fluentui/theme/lib/fonts/DefaultFontStyles.js
import { LocalizedFontNames as r, LocalizedFontFamilies as n, FontWeights as s } from "./FluentFonts.js";
import { createFontStyles as d } from "./createFontStyles.js";
import { getLanguage as l } from "../../../utilities/lib/language.js";
import { getWindow as f } from "../../../utilities/lib/dom/getWindow.js";
import { fontFace as w } from "../../../merge-styles/lib/fontFace.js";
var h = "https://res-1.cdn.office.net/files/fabric-cdn-prod_20230815.002/assets", W = /* @__PURE__ */ d(l());
function g(i, e, t, u) {
  i = "'".concat(i, "'");
  var a = u !== void 0 ? "local('".concat(u, "'),") : "";
  w({
    fontFamily: i,
    src: a + "url('".concat(e, ".woff2') format('woff2'),") + "url('".concat(e, ".woff') format('woff')"),
    fontWeight: t,
    fontStyle: "normal",
    fontDisplay: "swap"
  });
}
function o(i, e, t, u, a) {
  u === void 0 && (u = "segoeui");
  var c = "".concat(i, "/").concat(t, "/").concat(u);
  g(e, c + "-light", s.light, a && a + " Light"), g(e, c + "-semilight", s.semilight, a && a + " SemiLight"), g(e, c + "-regular", s.regular, a), g(e, c + "-semibold", s.semibold, a && a + " SemiBold"), g(e, c + "-bold", s.bold, a && a + " Bold");
}
function b(i) {
  if (i) {
    var e = "".concat(i, "/fonts");
    o(e, r.Thai, "leelawadeeui-thai", "leelawadeeui"), o(e, r.Arabic, "segoeui-arabic"), o(e, r.Cyrillic, "segoeui-cyrillic"), o(e, r.EastEuropean, "segoeui-easteuropean"), o(e, r.Greek, "segoeui-greek"), o(e, r.Hebrew, "segoeui-hebrew"), o(e, r.Vietnamese, "segoeui-vietnamese"), o(e, r.WestEuropean, "segoeui-westeuropean", "segoeui", "Segoe UI"), o(e, n.Selawik, "selawik", "selawik"), o(e, r.Armenian, "segoeui-armenian"), o(e, r.Georgian, "segoeui-georgian"), g("Leelawadee UI Web", "".concat(e, "/leelawadeeui-thai/leelawadeeui-semilight"), s.light), g("Leelawadee UI Web", "".concat(e, "/leelawadeeui-thai/leelawadeeui-bold"), s.semibold);
  }
}
function p() {
  var i, e, t = (i = f()) === null || i === void 0 ? void 0 : i.FabricConfig;
  return (e = t == null ? void 0 : t.fontBaseUrl) !== null && e !== void 0 ? e : h;
}
p();
export {
  W as DefaultFontStyles,
  b as registerDefaultFontFaces
};
