// @fluentui/react/lib/components/Image/Image.styles.js
import { getGlobalClassNames as x } from "../../../../style-utilities/lib/styles/getGlobalClassNames.js";
import "../../../../style-utilities/lib/index.js";
import { getWindow as L } from "../../../../utilities/lib/dom/getWindow.js";
import { AnimationClassNames as y } from "../../../../style-utilities/lib/classNames/AnimationClassNames.js";
var z = {
  root: "ms-Image",
  rootMaximizeFrame: "ms-Image--maximizeFrame",
  image: "ms-Image-image",
  imageCenter: "ms-Image-image--center",
  imageContain: "ms-Image-image--contain",
  imageCover: "ms-Image-image--cover",
  imageCenterContain: "ms-Image-image--centerContain",
  imageCenterCover: "ms-Image-image--centerCover",
  imageNone: "ms-Image-image--none",
  imageLandscape: "ms-Image-image--landscape",
  imagePortrait: "ms-Image-image--portrait"
}, W = function(e) {
  var I = e.className, o = e.width, n = e.height, w = e.maximizeFrame, r = e.isLoaded, h = e.shouldFadeIn, f = e.shouldStartVisible, a = e.isLandscape, d = e.isCenter, s = e.isContain, g = e.isCover, C = e.isCenterContain, l = e.isCenterCover, F = e.isNone, N = e.isError, b = e.isNotImageFit, c = e.theme, i = x(z, c), m = {
    position: "absolute",
    left: "50% /* @noflip */",
    top: "50%",
    transform: "translate(-50%,-50%)"
    // @todo test RTL renders transform: translate(50%,-50%);
  }, v = L(), t = v !== void 0 && // eslint-disable-next-line @fluentui/max-len
  // cast needed as vendor prefixed `msMaxTouchPoints` api is no longer part of TS lib declaration - introduced with TS 4.4
  v.navigator.msMaxTouchPoints === void 0, u = s && a || g && !a ? { width: "100%", height: "auto" } : { width: "auto", height: "100%" };
  return {
    root: [
      i.root,
      c.fonts.medium,
      {
        overflow: "hidden"
      },
      w && [
        i.rootMaximizeFrame,
        {
          height: "100%",
          width: "100%"
        }
      ],
      r && h && !f && y.fadeIn400,
      (d || s || g || C || l) && {
        position: "relative"
      },
      I
    ],
    image: [
      i.image,
      {
        display: "block",
        opacity: 0
      },
      r && [
        "is-loaded",
        {
          opacity: 1
        }
      ],
      d && [i.imageCenter, m],
      s && [
        i.imageContain,
        t && {
          width: "100%",
          height: "100%",
          objectFit: "contain"
        },
        !t && u,
        !t && m
      ],
      g && [
        i.imageCover,
        t && {
          width: "100%",
          height: "100%",
          objectFit: "cover"
        },
        !t && u,
        !t && m
      ],
      C && [
        i.imageCenterContain,
        a && {
          maxWidth: "100%"
        },
        !a && {
          maxHeight: "100%"
        },
        m
      ],
      l && [
        i.imageCenterCover,
        a && {
          maxHeight: "100%"
        },
        !a && {
          maxWidth: "100%"
        },
        m
      ],
      F && [
        i.imageNone,
        {
          width: "auto",
          height: "auto"
        }
      ],
      b && [
        !!o && !n && {
          height: "auto",
          width: "100%"
        },
        !o && !!n && {
          height: "100%",
          width: "auto"
        },
        !!o && !!n && {
          height: "100%",
          width: "100%"
        }
      ],
      a && i.imageLandscape,
      !a && i.imagePortrait,
      !r && "is-notLoaded",
      h && "is-fadeIn",
      N && "is-error"
    ]
  };
};
export {
  W as getStyles
};
